{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# FUNCTIONAL CHARACTERIZATION TABLE"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## RegulomeDB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax (<ipython-input-1-0f3d549778de>, line 7)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"<ipython-input-1-0f3d549778de>\"\u001b[0;36m, line \u001b[0;32m7\u001b[0m\n\u001b[0;31m    regulome <-  regulome %>% mutate(RegulomeImportance = case_when(\u001b[0m\n\u001b[0m                           ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid syntax\n"
     ]
    }
   ],
   "source": [
    "library (tidyr)\n",
    "library(dplyr)\n",
    "library(data.table)\n",
    "\n",
    "#Load the raw data\n",
    "regulome <- fread('/home/mentoraje/PAEs/SNProject/RegulomeDB/TSTFF344324.tsv', sep = '\\t')\n",
    "\n",
    "#Change the column name from score to Regulome Score for better understanding\n",
    "colnames(regulome)[12] <- 'RegulomeScore'\n",
    "\n",
    "#Add the numeric impact column called RegulomeImportance\n",
    "regulome <-  regulome %>% mutate(RegulomeImportance = case_when(\n",
    "  RegulomeScore %in% '1a' ~ 1,\n",
    "  RegulomeScore %in% '1b' ~ 2,\n",
    "  RegulomeScore %in% '1c' ~ 3,\n",
    "  RegulomeScore %in% '1d' ~ 4,\n",
    "  RegulomeScore %in% '1e' ~ 5,\n",
    "  RegulomeScore %in% '1f' ~ 6,\n",
    "  RegulomeScore %in% '2a' ~ 7,\n",
    "  RegulomeScore %in% '2b' ~ 8,\n",
    "  RegulomeScore %in% '2c' ~ 9,\n",
    "  RegulomeScore %in% '3a' ~ 10,\n",
    "  RegulomeScore %in% '3b' ~ 11,\n",
    "  RegulomeScore %in% '4' ~ 12,\n",
    "  RegulomeScore %in% '5' ~ 13,\n",
    "  RegulomeScore %in% '6' ~ 14,\n",
    "  RegulomeScore %in% '7' ~ 15\n",
    "  ))\n",
    "\n",
    "#For each cromosome write a file with the data. Only rsid, RegulomeScore and RegulomeImportance columns\n",
    "CHR <- c(1:22)\n",
    "for (chr in CHR){\n",
    "  print(paste0('-->',chr))\n",
    "  regulome_fltr <- regulome %>% select(rsid, RegulomeScore, RegulomeImportance) %>% filter(regulome$chrom == paste0('chr',chr))\n",
    "  \n",
    "  fwrite(regulome_fltr, file = paste0('/home/mentoraje/PAEs/SNProject/RegulomeDB/Resultats/RegulomeChr',chr,'.txt'), sep = '\\t')\n",
    "  \n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## ClinVar "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library (tidyr)\n",
    "library(dplyr)\n",
    "library(data.table)\n",
    "\n",
    "#Load the raw file\n",
    "clinvar <- fread('/home/mentoraje/PAEs/SNProject/ClinVar/variant_summary_new.txt', header = TRUE)\n",
    "\n",
    "#Change the column name to rsid for future merging\n",
    "colnames(clinvar)[10] <- 'rsid'\n",
    "\n",
    "#Remove duplicated SNPs, those that aren't from GRCh37 assembly and those that don't have rsid. \n",
    "clinvarGRch37 <- clinvar %>% \n",
    "  distinct(rsid, .keep_all = TRUE) %>% \n",
    "  filter(Assembly == 'GRCh37', rsid != -1)\n",
    "\n",
    "#Add the 'rs' before the id number for future merging\n",
    "clinvarGRch37$rsid <- c(paste0('rs',clinvarGRch37$rsid))\n",
    "\n",
    "#Select the columns we are interested in\n",
    "clinvar_fltr <- clinvarGRch37 %>% select(rsid, Chromosome, \n",
    "                                         ClinicalSignificance)\n",
    "\n",
    "#Separate the Clinical significance values, from the SNPs that have more than one, in different rows \n",
    "clinvar_fltr <- separate_rows(clinvar_fltr, ClinicalSignificance, sep = ', ')\n",
    "\n",
    "#Add a new column with the numeric impact (ClinicalImpact) of Clinical Significance \n",
    "clinvar_fltr <- clinvar_fltr %>% mutate(ClinicalImpact = case_when(\n",
    "  ClinicalSignificance %in% 'Pathogenic' ~ 1,\n",
    "  ClinicalSignificance %in% 'Pathogenic/Likely pathogenic' ~ 2,\n",
    "  ClinicalSignificance %in% 'Likely pathogenic' ~ 3,\n",
    "  ClinicalSignificance %in% 'drug response' ~ 4,\n",
    "  ClinicalSignificance %in% 'association' ~ 5, \n",
    "  ClinicalSignificance %in% 'risk factor' ~ 6,\n",
    "  ClinicalSignificance %in% 'Affects' ~ 7, \n",
    "  ClinicalSignificance %in% 'other' ~ 8,\n",
    "  ClinicalSignificance %in% 'Conflicting interpretations of pathogenicity' ~ 9,\n",
    "  ClinicalSignificance %in% 'Conflicting interpretations from submitters' ~ 10,\n",
    "  ClinicalSignificance %in% 'Uncertain significance' ~ 11,\n",
    "  ClinicalSignificance %in% 'no interpretation for the single variant' ~ 12,\n",
    "  ClinicalSignificance %in% 'not provided' ~ 13,\n",
    "  ClinicalSignificance %in% 'association not found' ~ 14,\n",
    "  ClinicalSignificance %in% 'protective' ~ 15,\n",
    "  ClinicalSignificance %in% 'Likely benign' ~ 16,\n",
    "  ClinicalSignificance %in% 'Benign/Likely benign' ~ 17,\n",
    "  ClinicalSignificance %in% 'Benign' ~ 18,\n",
    "  \n",
    "))\n",
    "\n",
    "#Regroup the rows by rsid and create a new column with maximum Clinical Impact (most severe)\n",
    "clinvar_def <- clinvar_fltr %>%\n",
    "  group_by(rsid) %>%\n",
    "  summarise(Chr = max(Chromosome),\n",
    "            ClinicalSignificances = toString(ClinicalSignificance),\n",
    "            ClinicaImpacts = toString(ClinicalImpact),\n",
    "            MaxClinicalImpact = min(ClinicalImpact)) %>%\n",
    "  ungroup() \n",
    "\n",
    "CHR <- c(1:22, 'X')\n",
    "\n",
    "#Write the data in a different file for each chromosome\n",
    "for (chr in CHR){\n",
    "  print(paste0('-->',chr))\n",
    "  clinvar_chr <- clinvar_def %>% filter(Chr == chr) %>% select(-Chr)\n",
    "  \n",
    "  if (length(pull(clinvar_chr, rsid)) == n_distinct(select(clinvar_chr, rsid))) {\n",
    "    fwrite(clinvar_chr, file = paste0('/home/mentoraje/PAEs/SNProject/ClinVar/Resultats/ClinVarChr',chr,'.txt'), sep = '\\t', dec ='.')\n",
    "    print('OK')\n",
    "  }\n",
    "  \n",
    "  else{print('NOT OK')} \n",
    "  \n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## DisGeNET "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library (tidyr)\n",
    "library(dplyr)\n",
    "library(data.table)\n",
    "\n",
    "#Load the raw data\n",
    "dades <- fread('/home/mentoraje/PAEs/SNProject/Disgenet/all_variant_disease_associations.tsv')\n",
    "\n",
    "CHR <- c(1:22)\n",
    "\n",
    "#For each chromosome\n",
    "for (chr in CHR){\n",
    "  print(paste(\"-->\", chr, sep=\"\"))\n",
    "  \n",
    "  #Change the column name for future merging\n",
    "  colnames(dades)[1] <- 'rsid'\n",
    "  \n",
    "  #Select the columns we are interested in and filter the SNPs by chromosome \n",
    "  dades_fltr <- dades %>% filter(dades$chromosome == chr) %>% distinct(rsid, .keep_all = TRUE) %>% select(rsid, chromosome, DSI, DPI, score, NofPmids) \n",
    "  \n",
    "  #Summarize the values from repeated SNPs (are associated to more than 1 disease) keeping the maximum value from each column \n",
    "  summ <- dades_fltr %>% group_by(rsid) %>% summarise( max(DSI), \n",
    "                                                       max(DPI),\n",
    "                                                       max(score), \n",
    "                                                       max(NofPmids))\n",
    "  #Write the data in a different file for each chromosome\n",
    "  fwrite(summ, file = paste0('/home/mentoraje/PAEs/SNProject/Disgenet/Resultats/chr',chr,'.txt'), sep = '\\t', dec ='.')\n",
    "  \n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Atlas of Variant Age "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library (tidyr)\n",
    "library(dplyr)\n",
    "library(data.table)\n",
    "\n",
    "CHR <- c(1:22)\n",
    "\n",
    "#For each chromosome \n",
    "for (chr in CHR) {\n",
    "  print(paste0('-->', chr))\n",
    "  \n",
    "  #load the raw data\n",
    "  Age <- fread(file = paste0('/home/mentoraje/PAEs/SNProject/Atlas_Variant_Age/DATA/atlas.chr',chr,'.csv'), header = TRUE, sep = ',') \n",
    "  \n",
    "  #Change the column name to rsid for future merging \n",
    "  colnames(Age)[1] <- 'rsid'\n",
    "  \n",
    "  #Select those SNPs from the 1000GP source and the columns we are interested in\n",
    "  Age_fltr <- Age %>% filter(Age$DataSource == 'TGP') %>% select(rsid, Position, AlleleAnc, starts_with('Age'), starts_with('Qual')) \n",
    "  \n",
    "  #Write the data in a different file for each chromosome \n",
    "  fwrite(Age_fltr, file = paste0('/home/mentoraje/PAEs/SNProject/Atlas_Variant_Age/Resultats/chr',chr,'.txt'),sep = \"\\t\")\n",
    "  \n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## MERGE DisGeNet, ClinVar and RegulomeDB "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library (tidyr)\n",
    "library(dplyr)\n",
    "library(data.table)\n",
    "\n",
    "\n",
    "CHR <- c(1:22)\n",
    "#For each chromosome\n",
    "for (chr in CHR){\n",
    "  print(paste0('-->',chr))\n",
    "  \n",
    "  #Load the filtered files from DisGeNET, ClinVar and RegulomeDB\n",
    "  disgenet <- fread(paste0('/home/mentoraje/PAEs/SNProject/Disgenet/Resultats/chr',chr,'.txt'))\n",
    "  clinvar <- fread(paste0('/home/mentoraje/PAEs/SNProject/ClinVar/Resultats/ClinVarChr',chr,'.txt'))\n",
    "  regulome <- fread(paste0('/home/mentoraje/PAEs/SNProject/RegulomeDB/Resultats/RegulomeChr',chr,'.txt'))\n",
    "\n",
    "  #Merge the three databases using full join (keep all the SNPs)   \n",
    "  merge1 <- full_join(clinvar, regulome)\n",
    "  merge2 <- full_join(merge1, disgenet)\n",
    "  \n",
    "  #Check the merge is correct\n",
    "  rsid_disgenet <- pull(disgenet, rsid) \n",
    "  rsid_regulome <- pull(regulome, rsid)\n",
    "  rsid_clinvar <- pull(clinvar, rsid)\n",
    "  \n",
    "  rsid_merge <- c(rsid_disgenet, rsid_regulome, rsid_clinvar)\n",
    "  \n",
    "  #If it is correct write a file per chromosome \n",
    "  if (n_distinct(rsid_merge) == length(pull(merge2, rsid))){\n",
    "    print('OK')\n",
    "    fwrite(merge2, file = paste0('/home/mentoraje/PAEs/SNProject/Merges/Merge_def/Resultats/MergeChr',chr,'.txt'), sep = '\\t', dec ='.')\n",
    "    \n",
    "  }\n",
    "  else{print('Not OK')}\n",
    "   \n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## MERGE ALL "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library (tidyr)\n",
    "library(dplyr)\n",
    "library(data.table)\n",
    "\n",
    "CHR <- c(1:22)\n",
    "for (chr in CHR){\n",
    "  \n",
    "  #Load the data from the previous merge (afegir), filtered age database (age) and 1000GP data with iHS, VEP and GWAS information (taula), previously merged by Daniel Soto \n",
    "  taula <- fread(paste0('/home/mentoraje/PAEs/SNProject/VEP/SNP_VEP_CHR',chr,'.txt'))\n",
    "  afegir <- fread(paste0('/home/mentoraje/PAEs/SNProject/Merges/Merge_taula_general/Resultats_merge_1/MergeChr',chr,'.txt'), sep = '\\t', dec ='.')\n",
    "  age  <- fread(paste0('/home/mentoraje/PAEs/SNProject/Merges/Merge_taula_general/Resultats_age/chr',chr,'.txt'))\n",
    "  \n",
    "  #Merge 1000GP (+ iHS, GWAS, VEP) data with previously merged data\n",
    "  taula_afegit <- left_join(taula, afegir)\n",
    "  \n",
    "  #Change the column name to physicalPos for future merging\n",
    "  colnames(age)[2] <- 'physicalPos'\n",
    "  \n",
    "  #Remove the rsid column from age data to merge only by position\n",
    "  age_sense_rsid <- age %>% select(!rsid)\n",
    "  \n",
    "  #Merge the previously merged data with age data\n",
    "  taula_junt <- left_join(taula_afegit, age_sense_rsid)\n",
    "  \n",
    "  ##ADD MEAN, MEDIAN AND SD FROM iHS COLUMNS \n",
    "  #Select only the iHS columns\n",
    "  dades_ihs <- taula_junt %>% select(starts_with('standiHS_'))\n",
    "  \n",
    "  #Convert the data to matrix format\n",
    "  dades_ihs <- as.matrix(dades_ihs)\n",
    "    \n",
    "  #Calculate mean, median and sd for each row\n",
    "  mean_calc <- apply(dades_ihs, 1, mean, na.rm = T)\n",
    "  median_calc <- apply(dades_ihs, 1, median, na.rm = T)\n",
    "  sd_calc <- apply(dades_ihs, 1, sd, na.rm = T)\n",
    "  \n",
    "  #Add the mean, median and sd columns\n",
    "  dades_mean <- add_column(taula_junt, iHS_mean = mean_calc, .after = 'chr')\n",
    "  dades_median <- add_column(dades_mean, iHS_median = median_calc, .after = 'iHS_mean')\n",
    "  dades_final <- add_column(dades_median, iHS_sd = sd_calc, .after = 'iHS_median')\n",
    "  \n",
    "  #Write the data in a file for each chromosome\n",
    "  fwrite(dades_final, paste0('/home/mentoraje/PAEs/SNProject/Merges/Merge_taula_general/Resultats/Taula_def_chr_',chr,'.txt'),  sep = '\\t', dec ='.')\n",
    "\n",
    "}\n",
    "\n",
    "##Merge all the chromosomes \n",
    "#load the final data from chromosome 1\n",
    "taula_inicial <- fread('/home/mentoraje/PAEs/SNProject/Merges/Merge_taula_general/Resultats/Taula_def_chr_1.txt')\n",
    "ADD <- c(2:22)\n",
    "#for each chromosome (but chr1) load the data from the chromosome variants and add them to the final table\n",
    "for (x in ADD){\n",
    "  taula_afegir <- fread(paste0('/home/mentoraje/PAEs/SNProject/Merges/Merge_taula_general/Resultats/Taula_def_chr_',chr,'.txt'))\n",
    "  taula_inicial <- bind_rows(taula_inicial, taula_afegir)\n",
    "}\n",
    "\n",
    "#Write the final data in a file\n",
    "fwrite(taula_inicial, '/home/mentoraje/PAEs/SNProject/Merges/Merge_taula_general/Resultats/TAULA_DEFINITIVA.txt')\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
