{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# SHINY APP"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Code for the creation of the interactive tool made with Shiny (from RStudios)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(ggplot2)\n",
    "library(shiny)\n",
    "library(shinyWidgets)\n",
    "library(dplyr)\n",
    "library(shinyjs) \n",
    "library(tidyverse)\n",
    "library(data.table)\n",
    "library(magrittr)\n",
    "library(shinyhelper)\n",
    "\n",
    "#Upload the table with all the information\n",
    "dades <- fread('TaulaExemple_sum.txt',  sep = '\\t', dec ='.', header = T) \n",
    "\n",
    "llista_rsid <- c('')\n",
    "\n",
    "#Vector of all populations from the 1000GP\n",
    "pops <- c('ACB', 'ASW', 'BEB', 'CDX', 'CEU', 'CHB', 'CHS', 'ESN', \n",
    "          'FIN', 'GBR', 'GIH', 'GWD', 'IBS', 'ITU', 'JPT', 'KHV', \n",
    "          'LWK', 'MSL', 'PJL', 'STU', 'TSI', 'YRI')\n",
    "\n",
    "#Vector with all the VEP consequences\n",
    "VEPLabels <- c(\"transcript_ablation\",\"splice_acceptor_variant\",\"splice_donor_variant\",\"stop_gained\",\n",
    "               \"frameshift_variant\",\"stop_lost\",\"start_lost\",\"transcript_amplification\",\n",
    "               \"inframe_insertion\",\"inframe_deletion\",\"missense_variant\",\"protein_altering_variant\",\n",
    "               \"splice_region_variant\",\"incomplete_terminal_codon_variant\",\"start_retained_variant\",\n",
    "               \"stop_retained_variant\",\"synonymous_variant\",\"coding_sequence_variant\",\n",
    "               \"mature_miRNA_variant\",\"5_prime_UTR_variant\",\"3_prime_UTR_variant\",\n",
    "               \"non_coding_transcript_exon_variant\",\"intron_variant\",\"NMD_transcript_variant\",\n",
    "               \"non_coding_transcript_variant\",\"upstream_gene_variant\",\"downstream_gene_variant\",\n",
    "               \"TFBS_ablation\",\"TFBS_amplification\",\"TF_binding_site_variant\",\n",
    "               \"regulatory_region_ablation\",\"regulatory_region_amplification\",\"feature_elongation\",\n",
    "               \"regulatory_region_variant\",\"feature_truncation\",\"intergenic_variant\")\n",
    "\n",
    "#Vector with all the Clinical Significances\n",
    "ClinVarLabels <- c('Pathogenic', 'Pathogenic/likely pathogenic', 'Likely pathogenic',\n",
    "                   'drug response', 'association', 'risk factor', 'affects', 'other',\n",
    "                   'conflicting interpretations of pathogenicity', \n",
    "                   'conflicting interpretations from submitters', 'uncertain significance',\n",
    "                   'no interpretation for the single variant', 'not provided', \n",
    "                   'association not found', 'protective', 'Likely benign', \n",
    "                   'Benign/Likely benign', 'Benign')\n",
    "\n",
    "#Vector with all the Regulome Rank Scores\n",
    "regulomeLabels <- c('1a', '1b', '1c', '1d', '1e', '1f', '2a',\n",
    "                    '2b', '2c', '3a', '3b', '4','5','6','7')\n",
    "\n",
    "\n",
    "\n",
    "ui <- fluidPage(\n",
    "\n",
    "    # Application title\n",
    "    titlePanel(\"Functional Description\"),\n",
    "    \n",
    "    sidebarLayout(\n",
    "        \n",
    "        #Sidebar panel (inputs)\n",
    "        sidebarPanel(\n",
    "            \n",
    "            #Filter title\n",
    "            h4('Location filters'),\n",
    "            \n",
    "            #Filters to select region\n",
    "            selectInput('chrInput', 'Chromosome:',\n",
    "                        choices = c(1:22),\n",
    "                        selected = 22),\n",
    "            \n",
    "            textInput('minPosInput', 'Min Position', value = min(dades$physicalPos)),\n",
    "            \n",
    "            textInput('maxPosInput', 'Max Position', value = max(dades$physicalPos)),\n",
    "\n",
    "\n",
    "            hr(), br(), #Separators\n",
    "            \n",
    "            #Filter title\n",
    "            h4('Population metrics filters'),\n",
    "            \n",
    "            #Filter by extreme values\n",
    "            checkboxInput('extremeValuesInput', 'Filter Extreme Values for IHS'),\n",
    "            \n",
    "            #Filter to select the populations where we apply the extreme values filter\n",
    "            checkboxGroupButtons('popsInput', 'Populations:', \n",
    "                               choices = pops, selected = pops),\n",
    "            \n",
    "            #Button to select/deselect all populations (all selected by default)\n",
    "            actionButton('selectAllPopsInput', 'Select / Deselect All'), \n",
    "            \n",
    "            hr(), br(), #Separators\n",
    "            \n",
    "            #Filter title\n",
    "            h4('Genome-wide Association Studies (GWAS) Catalog filter'),\n",
    "            \n",
    "            #Filter by GWAS association counts + help button with info about the filter\n",
    "            checkboxInput('filterGWASInput','FIlter by GWAS Association Counts' )%>%\n",
    "            helper(type = \"inline\", \n",
    "                   title = \"GWAS Filter Help\",\n",
    "                   content = c(\"GWAS Association counts filters the genome variants by how many times each variant has been reported in a GWAS study.\",\n",
    "                               \"GWAS catalog: <a>https://www.ebi.ac.uk/gwas</a>\"),\n",
    "                   size = \"m\"),\n",
    "            \n",
    "            #Slider to select the range of GWAS associations\n",
    "            sliderInput('GWASInput', 'GWAS Association Counts:',\n",
    "                        min = 0, max = max(dades$GWASAssocCounts),\n",
    "                        value = c(0, max), step = 1), \n",
    "\n",
    "            hr(), br(), #Separators\n",
    "            \n",
    "            #Filter title\n",
    "            h4('Variant Effect Predictor (VEP) filter'),\n",
    "            \n",
    "            #Filter by VEP consequences + help button with more info about the filter\n",
    "            checkboxInput('filterVEPInput', 'Filter by VEP Consequences') %>%\n",
    "                helper(type = \"inline\",\n",
    "                       title = \"VEP Filter Help\",\n",
    "                       content = c(\"VEP Consequences filters the genome variants by \n",
    "                                   the effect each allele may have on each transcript.\",\n",
    "                                   'The consequence terms are defined by Sequence Ontology',\n",
    "                                   \"The description of each consequence can be found here:\", \n",
    "                                   \"<a>https://m.ensembl.org/info/genome/variation/prediction/predicted_data.html</a>\",\n",
    "                                   'VEP Webpage: <a>https://www.ensembl.org/info/docs/tools/vep/index.html</a>'),\n",
    "                       size = \"m\"),\n",
    "            \n",
    "            #Checkboxes to select which VEP Consequences filter for\n",
    "            checkboxGroupInput('VEPInput', 'VEP Consequences',\n",
    "                                 choices =  VEPLabels),\n",
    "            \n",
    "            #Button to select/deselect all\n",
    "            actionButton('selectAllVEPInput', 'Select / Deselect All'),\n",
    "            \n",
    "            hr(), br(), #Separator\n",
    "            \n",
    "            #Filter title\n",
    "            h4('ClinVar filter'),\n",
    "            \n",
    "            #Filter by Clinical Significances + help button with more info about the filter\n",
    "            checkboxInput('filterClinVarInput', 'Filter by Clinical Significances') %>%\n",
    "                helper(type = \"inline\",\n",
    "                       title = \"ClinVar Filter Help\",\n",
    "                       content = c(\"Clinical Significance filters the genome variants by the interpretation of the relationship of a variation to a condition, these interpretations are provided only by submitters.\",\n",
    "                                   'Representation of Clinical Significances can be found here:', \n",
    "                                   '<a>https://www.ncbi.nlm.nih.gov/clinvar/docs/clinsig</a>',\n",
    "                                   \"ClinVar Webpage: <a>https://www.ncbi.nlm.nih.gov/clinvar</a>\"),\n",
    "                       size = \"m\"),\n",
    "            \n",
    "            #Checkboxes to select for which Clinical Significances filter\n",
    "            checkboxGroupInput('ClinVarInput', 'Clinical Significances:',\n",
    "                               choices = ClinVarLabels),\n",
    "            \n",
    "            #Select/deselect all button\n",
    "            actionButton('selectAllClinVarInput', 'Select / Deselect All'),\n",
    "            \n",
    "            hr(), br(),  #Separator\n",
    "            \n",
    "            #Filter tilte\n",
    "            h4('RegulomeDB filter'),\n",
    "            \n",
    "            #Filter by RegulomeDB rank Score + help button with more info about the filter\n",
    "            checkboxInput('filterRegulomeInput', 'Filter by RegulomeDB Rank Score') %>%\n",
    "                helper(type = \"inline\",\n",
    "                       title = \"RegulomeDB Filter Help\",\n",
    "                       content = c(\"Regulome rank filters the genome variants by putative regulatory potential.\",\n",
    "                                   'Representation of RegulomeDB Ranking can be found here: <a>https://regulomedb.org/regulome-help</a>',\n",
    "                                   \"RegulomeDB Webpage: <a>https://regulomedb.org/regulome-search</a>\"),\n",
    "                       size = \"m\"),\n",
    "            \n",
    "            #Slider to select the range of values of Regulome Rank Score \n",
    "            sliderTextInput('RegulomeInput', 'Regulome Score:',\n",
    "                            choices = regulomeLabels, selected = range(regulomeLabels),\n",
    "                            grid = TRUE),\n",
    "            \n",
    "            hr(), br(), #Separator\n",
    "            \n",
    "            #Filter title + help button with general info about the database\n",
    "            h4('DisGeNET filters')%>%\n",
    "                helper(type = \"inline\",\n",
    "                       title = \"DisGeNET Filter Help\",\n",
    "                       content = c(\"Each DisGeNET specific filter is explained below.\",\n",
    "                                   \"DisGeNET Webpage: <a>https://www.disgenet.org/home</a>\"),\n",
    "                       size = \"m\"),\n",
    "            \n",
    "            #Filter by DSI + help button with info about the filter\n",
    "            checkboxInput('filterDSIInput', 'Filter by Disease Specificity Index') %>%\n",
    "                helper(type = \"inline\",\n",
    "                       title = \"Disease Specificity Index Filter Help\",\n",
    "                       content = c(\"Disease Specificity Index (DSI) reflects if a variant (or gene) is associated to several or fewer diseases.\",\n",
    "                                   \"The DSI ranges from 0.25 to 1. For example: TNF, associated to more than 1,500 diseases has a DSI of 0.263, while HCN2 is associated to one disease has a DSI of 1.\",\n",
    "                                   \"DisGeNet Metrics Info: <a>https://www.disgenet.org/dbinfo</a>\"),\n",
    "                       size = \"m\"),\n",
    "            \n",
    "            #Slider to choose the range of DSI values to filter for\n",
    "            sliderInput('DSIInput', 'Disease Specificity Index:', \n",
    "                         min = 0.25, max = 1, \n",
    "                         value = c(0,1)),\n",
    "            \n",
    "            #Filter by DPI + help button with info about the filter\n",
    "            checkboxInput('filterDPIInput', 'Filter by Disease Pleiotropy Index') %>%\n",
    "            helper(type = \"inline\",\n",
    "                   title = \"Disease Pleiotropy Index Filter Help\",\n",
    "                   content = c(\"Disease Pleiotropy Index (DPI) is similar to DSI but it considers if the multiple diseases associated to the variant (or gene) are similar among them or are completely different diseases and belong to different disease clases.\",\n",
    "                               \"DisGeNet Metrics Info: <a>https://www.disgenet.org/dbinfo</a>\"),\n",
    "                   size = \"m\"),\n",
    "            \n",
    "            #Slider to choose the range of DPI values to filter for\n",
    "            sliderInput('DPIInput', 'Disease Pleiotropy Index:', \n",
    "                        min = 0, max = 1, \n",
    "                        value = c(0,1)),\n",
    "            \n",
    "            #Checkbox to filter by DisGeNET score + help button with info about the filter\n",
    "            checkboxInput('filterScoreInput', 'Filter by Variant-Disease Association Score') %>%\n",
    "            helper(type = \"inline\",\n",
    "                   title = \"DisGeNET Score Filter Help\",\n",
    "                   content = c(\"DisGeNET Score ranks the variant-disease association according to their level of evidence.\",\n",
    "                               \"DiSeNet Metrics Info: <a>https://www.disgenet.org/dbinfo</a>\"),\n",
    "                   size = \"m\"),\n",
    "            \n",
    "            #Slider to choose the range of score values to filter for\n",
    "            sliderInput('DGNscoreInput', 'Variant-Disease Association Score:', \n",
    "                        min = 0, max = 1, \n",
    "                        value = c(0,1)),\n",
    "            \n",
    "            #Filter by N.of Pubmed ids\n",
    "            checkboxInput('filterNofPmidsInput', 'Filter by Number of Pubmed IDs'),\n",
    "            \n",
    "            #Slider to choose Number of Pubmed IDs values to filter for\n",
    "            sliderInput('NofPmidsInput', 'Number of Pubmed IDs:', \n",
    "                        min = 0, max = max(dades$`max(NofPmids)`, na.rm = TRUE), \n",
    "                        value = c(0, max)),\n",
    "           \n",
    "            hr(), br(),  #Separator  \n",
    "            \n",
    "            #Filter title\n",
    "            h4('Atlas of Variant Age filters'), \n",
    "            \n",
    "            #Filter by age mean and Quality score\n",
    "            checkboxInput('filterAgeInput', 'Filter by Age Mean and Quality Score'),\n",
    "            \n",
    "            #Select by which clock model filter the age mean + help button\n",
    "            radioButtons('AgeModelInput', 'Atlas of Variant Age Clock Model:',\n",
    "                         choices = c('Mut', 'Rec', 'Jnt'), \n",
    "                         selected = 'Jnt') %>%\n",
    "                helper(type = \"inline\",\n",
    "                       title = \"Clock Model Filter Help\",\n",
    "                       content = c(\"The Genealogical Estimation of Variant Age can be based on three different\n",
    "                                   coalescent-base clock models (Mutation, Recombination and Joint Clock).\",\n",
    "                                   \"More information on the models can be found here: \n",
    "                                   <a>https://human.genome.dating/info/about</a>\"),\n",
    "                       size = \"m\"),\n",
    "            \n",
    "            #Slider to choose the range of Age mean value to filter for + help button\n",
    "            sliderInput('AgeMeanInput', 'Age Mean (years ago):', \n",
    "                        min = min(dades$AgeMean_Jnt, na.rm = TRUE),\n",
    "                        max = max(dades$AgeMean_Jnt, na.rm = TRUE), \n",
    "                        value = c(min, max)) %>%\n",
    "                helper(type = \"inline\",\n",
    "                       title = \"Age Mean Filter Help\",\n",
    "                       content = c(\"Age mean filters those genomic variants that have an age mean estimation\n",
    "                                   of derived allele in the selected range of values.\",\n",
    "                                   \"Human Genome Dating webpage: \n",
    "                                   <a>https://human.genome.dating/info/about</a>\"),\n",
    "                       size = \"m\"),\n",
    "            #Slider to choose the range of score value to filter for\n",
    "            sliderInput('qualScoreInput', 'Quality Score:',\n",
    "                        min = 0, max = 1, value = c(0,1)),\n",
    "            \n",
    "            hr(), #separator\n",
    "            \n",
    "            #Button to download the filtered table\n",
    "            downloadBttn('downloadInput', 'Download filtered Datatable'),\n",
    "           \n",
    "            #Color black the separator line          \n",
    "            tags$head(\n",
    "                tags$style(HTML(\"hr {border-top: 1px solid #000000;}\"))\n",
    "            )\n",
    "        ),\n",
    "\n",
    "        \n",
    "        #Main panel (ouptuts)\n",
    "        mainPanel(\n",
    "            \n",
    "            uiOutput('qresults'), br(), br(), #Textline with number of results found\n",
    "           \n",
    "            h3('iHS histogram', align = 'center'), #plot title\n",
    "            plotOutput('plotIHS'),  #Histogram of iHS values for all populations\n",
    "            p('iHS values from all populations are represented', align = 'center'), br(), br(), #Plot description\n",
    "            \n",
    "            h3('GWAS Association Counts', align = 'center'), #plot title\n",
    "            plotOutput('plotGWAS'), br(), br(), #Histogram of GWAS Association Counts\n",
    "            \n",
    "            h3('VEP Consequences', align = 'center'), #plot title\n",
    "            plotOutput('plotVEP'), br(), br(), #Histogram of VEP consequences\n",
    "            \n",
    "            h3('Clinical Significances from ClinVar', align = 'center'), #plot title\n",
    "            plotOutput('plotClinVar'), br(), br(), #Histogram of Clinical Significances\n",
    "            \n",
    "            h3('RegulomeDB Ranking', align = 'center'), #plot title\n",
    "            plotOutput('plotRegulome'), br(), br(), #Histogram of RegulomeDB rank\n",
    "            \n",
    "            h3('DisGeNET Histograms', align = 'center'), br(), \n",
    "            \n",
    "            h4('Disease Specificity Index'), #plot title\n",
    "            plotOutput('plotDSI'), br(), br(), #Histogram of DSI\n",
    "            \n",
    "            h4('Disease Pleiotropy Index'), #plot title\n",
    "            plotOutput('plotDPI'), br(), br(), #Histogram of DPI\n",
    "            \n",
    "            h4('DisGeNET Score'), #plot title\n",
    "            plotOutput('plotScore'), br(), br(), #Histogram of Disgenet Score\n",
    "            \n",
    "            h4('Number Of PubMed IDs'), #plot title\n",
    "            plotOutput('plotNofPmids'), br(), br(), #Histogram of Number of PubMed IDs\n",
    "            \n",
    "            h3('Atlas of Variant Age Histograms', align = 'center'), br(),\n",
    "            \n",
    "            h4('Age Mean by Mutation Clock'), #plot title\n",
    "            plotOutput('plotAgeMeanMut'), br(), br(), #Histogram of Age Mean by mutation clock model\n",
    "            \n",
    "            h4('Age Mean by Recombination Clock'), #plot title\n",
    "            plotOutput('plotAgeMeanRec'), br(), br(), #Histogram of Age Mean by recombination clock model\n",
    "            \n",
    "            h4('Age Mean by Joint Clock'), #plot title\n",
    "            plotOutput('plotAgeMeanJnt'), br(), br(), #Histogram of Age Mean by joint clock model\n",
    "            \n",
    "            h3('Table with the filtered results'), #Table title\n",
    "            dataTableOutput('tableResults') #Data table with the filtered results\n",
    "        \n",
    "        )\n",
    "    )\n",
    ")\n",
    "\n",
    "# Define server logic \n",
    "server <- function(input, output, session) {\n",
    "    \n",
    "    #Output for help buttons\n",
    "    observe_helpers(withMathJax = T)\n",
    "    \n",
    "    #Select or deselect all populations\n",
    "    observe({\n",
    "        if (input$selectAllPopsInput > 0){\n",
    "            if (input$selectAllPopsInput %% 2 == 0){\n",
    "                updateCheckboxGroupButtons(session, 'popsInput', selected = pops)\n",
    "            }\n",
    "            else {\n",
    "                updateCheckboxGroupButtons(session, 'popsInput', selected = '')\n",
    "            }\n",
    "        }\n",
    "    })\n",
    "    \n",
    "    #Select or deselect all VEP consequences\n",
    "    observe({\n",
    "        if (input$selectAllVEPInput > 0){\n",
    "            if (input$selectAllVEPInput %% 2 == 0){\n",
    "                updateCheckboxGroupInput(session, 'VEPInput', selected = '')\n",
    "            }\n",
    "            else {\n",
    "                updateCheckboxGroupInput(session, 'VEPInput', selected = VEPLabels)\n",
    "            }\n",
    "        }\n",
    "    })\n",
    "    \n",
    "    #Select or deselect all Clinical Significances\n",
    "    observe({\n",
    "        if (input$selectAllClinVarInput > 0){\n",
    "            if (input$selectAllClinVarInput %% 2 == 0){\n",
    "                updateCheckboxGroupInput(session, 'ClinVarInput', selected = '')\n",
    "            }\n",
    "            else {\n",
    "                updateCheckboxGroupInput(session, 'ClinVarInput', selected = ClinVarLabels)\n",
    "            }\n",
    "        }\n",
    "    })\n",
    "    \n",
    "    #Change minimum and maximum values from Age Mean slider depending on the clock model selected\n",
    "    observeEvent(input$AgeModelInput,{\n",
    "        updateSliderInput(session, 'AgeMeanInput',\n",
    "                          min = min(dades[[paste0('AgeMean_',input$AgeModelInput)]], na.rm = TRUE),\n",
    "                          max = max(dades[[paste0('AgeMean_',input$AgeModelInput)]], na.rm = TRUE),\n",
    "                          value = c(min,max))\n",
    "        \n",
    "    })\n",
    "   \n",
    "    \n",
    "    #TEXT OUTPUT WITH THE NUMBER OF RESULTS FOUND\n",
    "    output$qresults <- renderUI({\n",
    "        \n",
    "        #Filter the original data by position and save it in filtered\n",
    "        filtered <- dades %>% filter(chr %in% input$chrInput,\n",
    "                                     physicalPos >= input$minPosInput,\n",
    "                                     physicalPos <= input$maxPosInput)\n",
    "        \n",
    "        #if extreme values checkbox is selected, select those variants that have \n",
    "        #an extreme value in at least one of the populations selected\n",
    "        if (input$extremeValuesInput == TRUE){\n",
    "            for (POP in pops){\n",
    "                if (POP %in% input$popsInput){\n",
    "                    llista_rsid <- c(llista_rsid, \n",
    "                                     filtered %>% filter(!is.na(filtered[[paste0('extremevalue_',POP)]])) %>%\n",
    "                                         select(rsid))\n",
    "                        \n",
    "                    }\n",
    "            }\n",
    "                filtered <- filtered %>% filter(rsid %in% llista_rsid)      \n",
    "        }\n",
    "        \n",
    "        #If GWAS checkbox is selected, select those variants that have a GWAS Association Count \n",
    "        #value inside the chosen range \n",
    "        if (input$filterGWASInput == T){\n",
    "            filtered <- filtered %>% filter(GWASAssocCounts >= input$GWASInput[1],\n",
    "                                            GWASAssocCounts <= input$GWASInput[2])\n",
    "        }\n",
    "        \n",
    "        #If VEP checkbox is selected, select those variants with at least one of the VEP Consequences chosen\n",
    "        if (input$filterVEPInput == T){\n",
    "            separatedVEP <- separate_rows(filtered, VEPConseqs, sep = ',')\n",
    "            ids_filter_VEP <- separatedVEP %>% filter(VEPConseqs %in% input$VEPInput)\n",
    "            filtered <- filtered %>% filter (rsid %in% ids_filter_VEP$rsid)\n",
    "        }\n",
    "        \n",
    "        #If ClinVar checkbox is selected, select those variants with at least one of the Clinical Significances chosen\n",
    "        if (input$filterClinVarInput == T){\n",
    "            separatedCV <- separate_rows(filtered, ClinicalSignificances, sep = ', ')\n",
    "            ids_filter_CV <- separatedCV %>% filter(ClinicalSignificances %in% input$ClinVarInput)\n",
    "            filtered <- filtered %>% filter (rsid %in% ids_filter_CV$rsid)\n",
    "        }\n",
    "        \n",
    "        #If Regulome checkbox is selected, select those variants with a regulome rank score value\n",
    "        #inside the chosen range\n",
    "        if (input$filterRegulomeInput == T){\n",
    "            filtered <- filtered %>% filter(RegulomeImportance >= match(input$RegulomeInput[1], regulomeLabels),\n",
    "                                            RegulomeImportance <= match(input$RegulomeInput[2], regulomeLabels))\n",
    "        }\n",
    "        \n",
    "        #If DSI checbox is selected, select those variants with a DSI value inside the chosen range\n",
    "        if (input$filterDSIInput == T){\n",
    "            filtered <- filtered %>% filter (`max(DSI)` >= input$DSIInput[1],\n",
    "                                             `max(DSI)` <= input$DSIInput[2])\n",
    "        }\n",
    "        \n",
    "        #If DPI checbox is selected, select those variants with a DPI value inside the chosen range\n",
    "        if (input$filterDPIInput == T){\n",
    "            filtered <- filtered %>% filter (`max(DPI)` >= input$DPIInput[1],\n",
    "                                             `max(DPI)` <= input$DPIInput[2])\n",
    "        }\n",
    "        \n",
    "        #If score checbox is selected, select those variants with a DisGeNET score inside the chosen range\n",
    "        if (input$filterScoreInput == T){\n",
    "            filtered <- filtered %>% filter (`max(score)` >= input$DGNscoreInput[1],\n",
    "                                             `max(score)` <= input$DGNscoreInput[2])\n",
    "        }\n",
    "        \n",
    "        #If N.of PubMed IDs checbox is selected, select those variants with a number of \n",
    "        #PubMed IDs value inside the chosen range\n",
    "        if (input$filterNofPmidsInput == T){\n",
    "            filtered <- filtered %>% filter (`max(NofPmids)` >= input$NofPmidsInput[1],\n",
    "                                             `max(NofPmids)` <= input$NofPmidsInput[2])\n",
    "        }\n",
    "        \n",
    "        #If Age filter checbox is selected, select those variants with a Age mean and score value, \n",
    "        #of the chosen clock model, inside the chosen ranges\n",
    "        if (input$filterAgeInput == T){\n",
    "            if ('Mut' %in% input$AgeModelInput){\n",
    "                filtered <- filtered %>% filter(AgeMean_Mut >= input$AgeMeanInput[1],\n",
    "                                                AgeMean_Mut <= input$AgeMeanInput[2],\n",
    "                                                QualScore_Mut >= input$qualScoreInput[1],\n",
    "                                                QualScore_Mut <= input$qualScoreInput[2])\n",
    "            }\n",
    "            if ('Rec' %in% input$AgeModelInput){\n",
    "                filtered <- filtered %>% filter(AgeMean_Rec >= input$AgeMeanInput[1],\n",
    "                                                AgeMean_Rec <= input$AgeMeanInput[2],\n",
    "                                                QualScore_Rec >= input$qualScoreInput[1],\n",
    "                                                QualScore_Rec <= input$qualScoreInput[2])\n",
    "            }\n",
    "            if ('Jnt' %in% input$AgeModelInput){\n",
    "                filtered <- filtered %>% filter(AgeMean_Jnt >= input$AgeMeanInput[1],\n",
    "                                                AgeMean_Jnt <= input$AgeMeanInput[2],\n",
    "                                                QualScore_Jnt >= input$qualScoreInput[1],\n",
    "                                                QualScore_Jnt <= input$qualScoreInput[2])\n",
    "            }\n",
    "        }\n",
    "        \n",
    "        #Print a text line with how many results were found\n",
    "        return(paste(nrow(filtered), 'results were found.'))\n",
    "        \n",
    "    })\n",
    "    \n",
    "    #iHS PLOT \n",
    "    output$plotIHS <- renderPlot({\n",
    "        \n",
    "        #Filter the original data by position and save it in filtered\n",
    "        filtered <- dades %>% filter(chr %in% input$chrInput,\n",
    "                                     physicalPos >= input$minPosInput,\n",
    "                                     physicalPos <= input$maxPosInput)\n",
    "        \n",
    "        #if extreme values checkbox is selected, select those variants that have \n",
    "        #an extreme value in at least one of the populations selected\n",
    "        if (input$extremeValuesInput == TRUE){\n",
    "            for (POP in pops){\n",
    "                if (POP %in% input$popsInput){\n",
    "                    llista_rsid <- c(llista_rsid, \n",
    "                                     filtered %>% filter(!is.na(filtered[[paste0('extremevalue_',POP)]])) %>%\n",
    "                                         select(rsid))\n",
    "                        \n",
    "                    }\n",
    "            }\n",
    "                filtered <- filtered %>% filter(rsid %in% llista_rsid)      \n",
    "        }\n",
    "        \n",
    "        #If GWAS checkbox is selected, select those variants that have a GWAS Association Count \n",
    "        #value inside the chosen range \n",
    "        if (input$filterGWASInput == T){\n",
    "            filtered <- filtered %>% filter(GWASAssocCounts >= input$GWASInput[1],\n",
    "                                            GWASAssocCounts <= input$GWASInput[2])\n",
    "        }\n",
    "        \n",
    "        #If VEP checkbox is selected, select those variants with at least one of the VEP Consequences chosen\n",
    "        if (input$filterVEPInput == T){\n",
    "            separatedVEP <- separate_rows(filtered, VEPConseqs, sep = ',')\n",
    "            ids_filter_VEP <- separatedVEP %>% filter(VEPConseqs %in% input$VEPInput)\n",
    "            filtered <- filtered %>% filter (rsid %in% ids_filter_VEP$rsid)\n",
    "        }\n",
    "        \n",
    "        #If ClinVar checkbox is selected, select those variants with at least one of the Clinical Significances chosen\n",
    "        if (input$filterClinVarInput == T){\n",
    "            separatedCV <- separate_rows(filtered, ClinicalSignificances, sep = ', ')\n",
    "            ids_filter_CV <- separatedCV %>% filter(ClinicalSignificances %in% input$ClinVarInput)\n",
    "            filtered <- filtered %>% filter (rsid %in% ids_filter_CV$rsid)\n",
    "        }\n",
    "        \n",
    "        #If Regulome checkbox is selected, select those variants with a regulome rank score value\n",
    "        #inside the chosen range\n",
    "        if (input$filterRegulomeInput == T){\n",
    "            filtered <- filtered %>% filter(RegulomeImportance >= match(input$RegulomeInput[1], regulomeLabels),\n",
    "                                            RegulomeImportance <= match(input$RegulomeInput[2], regulomeLabels))\n",
    "        }\n",
    "        \n",
    "        #If DSI checbox is selected, select those variants with a DSI value inside the chosen range\n",
    "        if (input$filterDSIInput == T){\n",
    "            filtered <- filtered %>% filter (`max(DSI)` >= input$DSIInput[1],\n",
    "                                             `max(DSI)` <= input$DSIInput[2])\n",
    "        }\n",
    "        \n",
    "        #If DPI checbox is selected, select those variants with a DPI value inside the chosen range\n",
    "        if (input$filterDPIInput == T){\n",
    "            filtered <- filtered %>% filter (`max(DPI)` >= input$DPIInput[1],\n",
    "                                             `max(DPI)` <= input$DPIInput[2])\n",
    "        }\n",
    "        \n",
    "        #If score checbox is selected, select those variants with a DisGeNET score inside the chosen range\n",
    "        if (input$filterScoreInput == T){\n",
    "            filtered <- filtered %>% filter (`max(score)` >= input$DGNscoreInput[1],\n",
    "                                             `max(score)` <= input$DGNscoreInput[2])\n",
    "        }\n",
    "        \n",
    "        #If N.of PubMed IDs checbox is selected, select those variants with a number of \n",
    "        #PubMed IDs value inside the chosen range\n",
    "        if (input$filterNofPmidsInput == T){\n",
    "            filtered <- filtered %>% filter (`max(NofPmids)` >= input$NofPmidsInput[1],\n",
    "                                             `max(NofPmids)` <= input$NofPmidsInput[2])\n",
    "        }\n",
    "        \n",
    "        #If Age filter checbox is selected, select those variants with a Age mean and score value, \n",
    "        #of the chosen clock model, inside the chosen ranges\n",
    "        if (input$filterAgeInput == T){\n",
    "            if ('Mut' %in% input$AgeModelInput){\n",
    "                filtered <- filtered %>% filter(AgeMean_Mut >= input$AgeMeanInput[1],\n",
    "                                                AgeMean_Mut <= input$AgeMeanInput[2],\n",
    "                                                QualScore_Mut >= input$qualScoreInput[1],\n",
    "                                                QualScore_Mut <= input$qualScoreInput[2])\n",
    "            }\n",
    "            if ('Rec' %in% input$AgeModelInput){\n",
    "                filtered <- filtered %>% filter(AgeMean_Rec >= input$AgeMeanInput[1],\n",
    "                                                AgeMean_Rec <= input$AgeMeanInput[2],\n",
    "                                                QualScore_Rec >= input$qualScoreInput[1],\n",
    "                                                QualScore_Rec <= input$qualScoreInput[2])\n",
    "            }\n",
    "            if ('Jnt' %in% input$AgeModelInput){\n",
    "                filtered <- filtered %>% filter(AgeMean_Jnt >= input$AgeMeanInput[1],\n",
    "                                                AgeMean_Jnt <= input$AgeMeanInput[2],\n",
    "                                                QualScore_Jnt >= input$qualScoreInput[1],\n",
    "                                                QualScore_Jnt <= input$qualScoreInput[2])\n",
    "            }\n",
    "        }\n",
    "        \n",
    "        #Select the rsid and all iHS columns\n",
    "        filtered <- filtered %>% select(rsid, starts_with('standiHS_'))\n",
    "        #Transform the data from wide to long format \n",
    "        filtered_long <- melt(setDT(filtered), id.vars = \"rsid\", variable.name = \"iHS\")\n",
    "        \n",
    "        #Design the plot (x axis = iHS values; y axis = count)\n",
    "        o <- ggplot(filtered_long, aes(x=value, y=..count.., fill = ..count..)) + geom_histogram() \n",
    "        o + scale_fill_gradient(low='skyblue', high = 'purple')\n",
    "        \n",
    "    })\n",
    "    \n",
    "\n",
    "    \n",
    "    #GWAS ASSOCIATION COUNTS PLOT\n",
    "    output$plotGWAS <- renderPlot({\n",
    "        \n",
    "        #Filter the original data by position and save it in filtered\n",
    "        filtered <- dades %>% filter(chr %in% input$chrInput,\n",
    "                                     physicalPos >= input$minPosInput,\n",
    "                                     physicalPos <= input$maxPosInput)\n",
    "        \n",
    "        #if extreme values checkbox is selected, select those variants that have \n",
    "        #an extreme value in at least one of the populations selected\n",
    "        if (input$extremeValuesInput == TRUE){\n",
    "            for (POP in pops){\n",
    "                if (POP %in% input$popsInput){\n",
    "                    llista_rsid <- c(llista_rsid, \n",
    "                                     filtered %>% filter(!is.na(filtered[[paste0('extremevalue_',POP)]])) %>%\n",
    "                                         select(rsid))\n",
    "                        \n",
    "                    }\n",
    "            }\n",
    "                filtered <- filtered %>% filter(rsid %in% llista_rsid)      \n",
    "        }\n",
    "        \n",
    "        #If GWAS checkbox is selected, select those variants that have a GWAS Association Count \n",
    "        #value inside the chosen range \n",
    "        if (input$filterGWASInput == T){\n",
    "            filtered <- filtered %>% filter(GWASAssocCounts >= input$GWASInput[1],\n",
    "                                            GWASAssocCounts <= input$GWASInput[2])\n",
    "        }\n",
    "        \n",
    "        #If VEP checkbox is selected, select those variants with at least one of the VEP Consequences chosen\n",
    "        if (input$filterVEPInput == T){\n",
    "            separatedVEP <- separate_rows(filtered, VEPConseqs, sep = ',')\n",
    "            ids_filter_VEP <- separatedVEP %>% filter(VEPConseqs %in% input$VEPInput)\n",
    "            filtered <- filtered %>% filter (rsid %in% ids_filter_VEP$rsid)\n",
    "        }\n",
    "        \n",
    "        #If ClinVar checkbox is selected, select those variants with at least one of the Clinical Significances chosen\n",
    "        if (input$filterClinVarInput == T){\n",
    "            separatedCV <- separate_rows(filtered, ClinicalSignificances, sep = ', ')\n",
    "            ids_filter_CV <- separatedCV %>% filter(ClinicalSignificances %in% input$ClinVarInput)\n",
    "            filtered <- filtered %>% filter (rsid %in% ids_filter_CV$rsid)\n",
    "        }\n",
    "        \n",
    "        #If Regulome checkbox is selected, select those variants with a regulome rank score value\n",
    "        #inside the chosen range\n",
    "        if (input$filterRegulomeInput == T){\n",
    "            filtered <- filtered %>% filter(RegulomeImportance >= match(input$RegulomeInput[1], regulomeLabels),\n",
    "                                            RegulomeImportance <= match(input$RegulomeInput[2], regulomeLabels))\n",
    "        }\n",
    "        \n",
    "        #If DSI checbox is selected, select those variants with a DSI value inside the chosen range\n",
    "        if (input$filterDSIInput == T){\n",
    "            filtered <- filtered %>% filter (`max(DSI)` >= input$DSIInput[1],\n",
    "                                             `max(DSI)` <= input$DSIInput[2])\n",
    "        }\n",
    "        \n",
    "        #If DPI checbox is selected, select those variants with a DPI value inside the chosen range\n",
    "        if (input$filterDPIInput == T){\n",
    "            filtered <- filtered %>% filter (`max(DPI)` >= input$DPIInput[1],\n",
    "                                             `max(DPI)` <= input$DPIInput[2])\n",
    "        }\n",
    "        \n",
    "        #If score checbox is selected, select those variants with a DisGeNET score inside the chosen range\n",
    "        if (input$filterScoreInput == T){\n",
    "            filtered <- filtered %>% filter (`max(score)` >= input$DGNscoreInput[1],\n",
    "                                             `max(score)` <= input$DGNscoreInput[2])\n",
    "        }\n",
    "        \n",
    "        #If N.of PubMed IDs checbox is selected, select those variants with a number of \n",
    "        #PubMed IDs value inside the chosen range\n",
    "        if (input$filterNofPmidsInput == T){\n",
    "            filtered <- filtered %>% filter (`max(NofPmids)` >= input$NofPmidsInput[1],\n",
    "                                             `max(NofPmids)` <= input$NofPmidsInput[2])\n",
    "        }\n",
    "        \n",
    "        #If Age filter checbox is selected, select those variants with a Age mean and score value, \n",
    "        #of the chosen clock model, inside the chosen ranges\n",
    "        if (input$filterAgeInput == T){\n",
    "            if ('Mut' %in% input$AgeModelInput){\n",
    "                filtered <- filtered %>% filter(AgeMean_Mut >= input$AgeMeanInput[1],\n",
    "                                                AgeMean_Mut <= input$AgeMeanInput[2],\n",
    "                                                QualScore_Mut >= input$qualScoreInput[1],\n",
    "                                                QualScore_Mut <= input$qualScoreInput[2])\n",
    "            }\n",
    "            if ('Rec' %in% input$AgeModelInput){\n",
    "                filtered <- filtered %>% filter(AgeMean_Rec >= input$AgeMeanInput[1],\n",
    "                                                AgeMean_Rec <= input$AgeMeanInput[2],\n",
    "                                                QualScore_Rec >= input$qualScoreInput[1],\n",
    "                                                QualScore_Rec <= input$qualScoreInput[2])\n",
    "            }\n",
    "            if ('Jnt' %in% input$AgeModelInput){\n",
    "                filtered <- filtered %>% filter(AgeMean_Jnt >= input$AgeMeanInput[1],\n",
    "                                                AgeMean_Jnt <= input$AgeMeanInput[2],\n",
    "                                                QualScore_Jnt >= input$qualScoreInput[1],\n",
    "                                                QualScore_Jnt <= input$qualScoreInput[2])\n",
    "            }\n",
    "        }\n",
    "        \n",
    "        #Design the plot\n",
    "        o <- ggplot(filtered, aes(x=GWASAssocCounts, y=..count.., fill = ..count..)) + geom_histogram() \n",
    "        o + scale_fill_gradient(low='skyblue', high = 'purple')\n",
    "        \n",
    "    })\n",
    "   \n",
    "     #VEP CONSEQUENCES PLOT\n",
    "    output$plotVEP <- renderPlot({\n",
    "        \n",
    "        #Filter the original data by position and save it in filtered\n",
    "        filtered <- dades %>% filter(chr %in% input$chrInput,\n",
    "                                     physicalPos >= input$minPosInput,\n",
    "                                     physicalPos <= input$maxPosInput)\n",
    "        \n",
    "        #if extreme values checkbox is selected, select those variants that have \n",
    "        #an extreme value in at least one of the populations selected\n",
    "        if (input$extremeValuesInput == TRUE){\n",
    "            for (POP in pops){\n",
    "                if (POP %in% input$popsInput){\n",
    "                    llista_rsid <- c(llista_rsid, \n",
    "                                     filtered %>% filter(!is.na(filtered[[paste0('extremevalue_',POP)]])) %>%\n",
    "                                         select(rsid))\n",
    "                        \n",
    "                    }\n",
    "            }\n",
    "                filtered <- filtered %>% filter(rsid %in% llista_rsid)      \n",
    "        }\n",
    "        \n",
    "        #If GWAS checkbox is selected, select those variants that have a GWAS Association Count \n",
    "        #value inside the chosen range \n",
    "        if (input$filterGWASInput == T){\n",
    "            filtered <- filtered %>% filter(GWASAssocCounts >= input$GWASInput[1],\n",
    "                                            GWASAssocCounts <= input$GWASInput[2])\n",
    "        }\n",
    "        \n",
    "        #If VEP checkbox is selected, select those variants with at least one of the VEP Consequences chosen\n",
    "        if (input$filterVEPInput == T){\n",
    "            separatedVEP <- separate_rows(filtered, VEPConseqs, sep = ',')\n",
    "            ids_filter_VEP <- separatedVEP %>% filter(VEPConseqs %in% input$VEPInput)\n",
    "            filtered <- filtered %>% filter (rsid %in% ids_filter_VEP$rsid)\n",
    "        }\n",
    "        \n",
    "        #If ClinVar checkbox is selected, select those variants with at least one of the Clinical Significances chosen\n",
    "        if (input$filterClinVarInput == T){\n",
    "            separatedCV <- separate_rows(filtered, ClinicalSignificances, sep = ', ')\n",
    "            ids_filter_CV <- separatedCV %>% filter(ClinicalSignificances %in% input$ClinVarInput)\n",
    "            filtered <- filtered %>% filter (rsid %in% ids_filter_CV$rsid)\n",
    "        }\n",
    "        \n",
    "        #If Regulome checkbox is selected, select those variants with a regulome rank score value\n",
    "        #inside the chosen range\n",
    "        if (input$filterRegulomeInput == T){\n",
    "            filtered <- filtered %>% filter(RegulomeImportance >= match(input$RegulomeInput[1], regulomeLabels),\n",
    "                                            RegulomeImportance <= match(input$RegulomeInput[2], regulomeLabels))\n",
    "        }\n",
    "        \n",
    "        #If DSI checbox is selected, select those variants with a DSI value inside the chosen range\n",
    "        if (input$filterDSIInput == T){\n",
    "            filtered <- filtered %>% filter (`max(DSI)` >= input$DSIInput[1],\n",
    "                                             `max(DSI)` <= input$DSIInput[2])\n",
    "        }\n",
    "        \n",
    "        #If DPI checbox is selected, select those variants with a DPI value inside the chosen range\n",
    "        if (input$filterDPIInput == T){\n",
    "            filtered <- filtered %>% filter (`max(DPI)` >= input$DPIInput[1],\n",
    "                                             `max(DPI)` <= input$DPIInput[2])\n",
    "        }\n",
    "        \n",
    "        #If score checbox is selected, select those variants with a DisGeNET score inside the chosen range\n",
    "        if (input$filterScoreInput == T){\n",
    "            filtered <- filtered %>% filter (`max(score)` >= input$DGNscoreInput[1],\n",
    "                                             `max(score)` <= input$DGNscoreInput[2])\n",
    "        }\n",
    "        \n",
    "        #If N.of PubMed IDs checbox is selected, select those variants with a number of \n",
    "        #PubMed IDs value inside the chosen range\n",
    "        if (input$filterNofPmidsInput == T){\n",
    "            filtered <- filtered %>% filter (`max(NofPmids)` >= input$NofPmidsInput[1],\n",
    "                                             `max(NofPmids)` <= input$NofPmidsInput[2])\n",
    "        }\n",
    "        \n",
    "        #If Age filter checbox is selected, select those variants with a Age mean and score value, \n",
    "        #of the chosen clock model, inside the chosen ranges\n",
    "        if (input$filterAgeInput == T){\n",
    "            if ('Mut' %in% input$AgeModelInput){\n",
    "                filtered <- filtered %>% filter(AgeMean_Mut >= input$AgeMeanInput[1],\n",
    "                                                AgeMean_Mut <= input$AgeMeanInput[2],\n",
    "                                                QualScore_Mut >= input$qualScoreInput[1],\n",
    "                                                QualScore_Mut <= input$qualScoreInput[2])\n",
    "            }\n",
    "            if ('Rec' %in% input$AgeModelInput){\n",
    "                filtered <- filtered %>% filter(AgeMean_Rec >= input$AgeMeanInput[1],\n",
    "                                                AgeMean_Rec <= input$AgeMeanInput[2],\n",
    "                                                QualScore_Rec >= input$qualScoreInput[1],\n",
    "                                                QualScore_Rec <= input$qualScoreInput[2])\n",
    "            }\n",
    "            if ('Jnt' %in% input$AgeModelInput){\n",
    "                filtered <- filtered %>% filter(AgeMean_Jnt >= input$AgeMeanInput[1],\n",
    "                                                AgeMean_Jnt <= input$AgeMeanInput[2],\n",
    "                                                QualScore_Jnt >= input$qualScoreInput[1],\n",
    "                                                QualScore_Jnt <= input$qualScoreInput[2])\n",
    "            }\n",
    "        }\n",
    "        \n",
    "        #separate the rows that have more than one VEP Consequence to count them independently \n",
    "        filter_separat <- separate_rows(filtered, VEPConseqs, sep = ',')\n",
    "        \n",
    "        #Design the plot \n",
    "        o <- ggplot(filter_separat, aes(x=VEPConseqs, stat=\"count\", fill = ..count..)) + geom_bar() \n",
    "        o + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + scale_fill_gradient(low = 'skyblue', high = 'purple')\n",
    "        \n",
    "    })\n",
    "    \n",
    "    #CLINVAR PLOT\n",
    "    output$plotClinVar <- renderPlot({\n",
    "        \n",
    "        #Filter the original data by position and save it in filtered\n",
    "        filtered <- dades %>% filter(chr %in% input$chrInput,\n",
    "                                     physicalPos >= input$minPosInput,\n",
    "                                     physicalPos <= input$maxPosInput)\n",
    "        \n",
    "        #if extreme values checkbox is selected, select those variants that have \n",
    "        #an extreme value in at least one of the populations selected\n",
    "        if (input$extremeValuesInput == TRUE){\n",
    "            for (POP in pops){\n",
    "                if (POP %in% input$popsInput){\n",
    "                    llista_rsid <- c(llista_rsid, \n",
    "                                     filtered %>% filter(!is.na(filtered[[paste0('extremevalue_',POP)]])) %>%\n",
    "                                         select(rsid))\n",
    "                        \n",
    "                    }\n",
    "            }\n",
    "                filtered <- filtered %>% filter(rsid %in% llista_rsid)      \n",
    "        }\n",
    "        \n",
    "        #If GWAS checkbox is selected, select those variants that have a GWAS Association Count \n",
    "        #value inside the chosen range \n",
    "        if (input$filterGWASInput == T){\n",
    "            filtered <- filtered %>% filter(GWASAssocCounts >= input$GWASInput[1],\n",
    "                                            GWASAssocCounts <= input$GWASInput[2])\n",
    "        }\n",
    "        \n",
    "        #If VEP checkbox is selected, select those variants with at least one of the VEP Consequences chosen\n",
    "        if (input$filterVEPInput == T){\n",
    "            separatedVEP <- separate_rows(filtered, VEPConseqs, sep = ',')\n",
    "            ids_filter_VEP <- separatedVEP %>% filter(VEPConseqs %in% input$VEPInput)\n",
    "            filtered <- filtered %>% filter (rsid %in% ids_filter_VEP$rsid)\n",
    "        }\n",
    "        \n",
    "        #If ClinVar checkbox is selected, select those variants with at least one of the Clinical Significances chosen\n",
    "        if (input$filterClinVarInput == T){\n",
    "            separatedCV <- separate_rows(filtered, ClinicalSignificances, sep = ', ')\n",
    "            ids_filter_CV <- separatedCV %>% filter(ClinicalSignificances %in% input$ClinVarInput)\n",
    "            filtered <- filtered %>% filter (rsid %in% ids_filter_CV$rsid)\n",
    "        }\n",
    "        \n",
    "        #If Regulome checkbox is selected, select those variants with a regulome rank score value\n",
    "        #inside the chosen range\n",
    "        if (input$filterRegulomeInput == T){\n",
    "            filtered <- filtered %>% filter(RegulomeImportance >= match(input$RegulomeInput[1], regulomeLabels),\n",
    "                                            RegulomeImportance <= match(input$RegulomeInput[2], regulomeLabels))\n",
    "        }\n",
    "        \n",
    "        #If DSI checbox is selected, select those variants with a DSI value inside the chosen range\n",
    "        if (input$filterDSIInput == T){\n",
    "            filtered <- filtered %>% filter (`max(DSI)` >= input$DSIInput[1],\n",
    "                                             `max(DSI)` <= input$DSIInput[2])\n",
    "        }\n",
    "        \n",
    "        #If DPI checbox is selected, select those variants with a DPI value inside the chosen range\n",
    "        if (input$filterDPIInput == T){\n",
    "            filtered <- filtered %>% filter (`max(DPI)` >= input$DPIInput[1],\n",
    "                                             `max(DPI)` <= input$DPIInput[2])\n",
    "        }\n",
    "        \n",
    "        #If score checbox is selected, select those variants with a DisGeNET score inside the chosen range\n",
    "        if (input$filterScoreInput == T){\n",
    "            filtered <- filtered %>% filter (`max(score)` >= input$DGNscoreInput[1],\n",
    "                                             `max(score)` <= input$DGNscoreInput[2])\n",
    "        }\n",
    "        \n",
    "        #If N.of PubMed IDs checbox is selected, select those variants with a number of \n",
    "        #PubMed IDs value inside the chosen range\n",
    "        if (input$filterNofPmidsInput == T){\n",
    "            filtered <- filtered %>% filter (`max(NofPmids)` >= input$NofPmidsInput[1],\n",
    "                                             `max(NofPmids)` <= input$NofPmidsInput[2])\n",
    "        }\n",
    "        \n",
    "        #If Age filter checbox is selected, select those variants with a Age mean and score value, \n",
    "        #of the chosen clock model, inside the chosen ranges\n",
    "        if (input$filterAgeInput == T){\n",
    "            if ('Mut' %in% input$AgeModelInput){\n",
    "                filtered <- filtered %>% filter(AgeMean_Mut >= input$AgeMeanInput[1],\n",
    "                                                AgeMean_Mut <= input$AgeMeanInput[2],\n",
    "                                                QualScore_Mut >= input$qualScoreInput[1],\n",
    "                                                QualScore_Mut <= input$qualScoreInput[2])\n",
    "            }\n",
    "            if ('Rec' %in% input$AgeModelInput){\n",
    "                filtered <- filtered %>% filter(AgeMean_Rec >= input$AgeMeanInput[1],\n",
    "                                                AgeMean_Rec <= input$AgeMeanInput[2],\n",
    "                                                QualScore_Rec >= input$qualScoreInput[1],\n",
    "                                                QualScore_Rec <= input$qualScoreInput[2])\n",
    "            }\n",
    "            if ('Jnt' %in% input$AgeModelInput){\n",
    "                filtered <- filtered %>% filter(AgeMean_Jnt >= input$AgeMeanInput[1],\n",
    "                                                AgeMean_Jnt <= input$AgeMeanInput[2],\n",
    "                                                QualScore_Jnt >= input$qualScoreInput[1],\n",
    "                                                QualScore_Jnt <= input$qualScoreInput[2])\n",
    "            }\n",
    "        }\n",
    "        \n",
    "        #Separate the rows that have more than one clinical significance to count them independently\n",
    "        filter_separat <- separate_rows(filtered, ClinicalSignificances, sep = ', ')\n",
    "       \n",
    "        #Design the plot\n",
    "        o <- ggplot(filter_separat, aes(x=ClinicalSignificances, stat=\"count\", fill = ..count..)) + geom_bar() \n",
    "        o + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + scale_fill_gradient(low = 'skyblue', high = 'purple')\n",
    "        \n",
    "    })\n",
    "    \n",
    "    #REGULOME PLOT\n",
    "    output$plotRegulome <- renderPlot({\n",
    "        \n",
    "        #Filter the original data by position and save it in filtered\n",
    "        filtered <- dades %>% filter(chr %in% input$chrInput,\n",
    "                                     physicalPos >= input$minPosInput,\n",
    "                                     physicalPos <= input$maxPosInput)\n",
    "        \n",
    "        #if extreme values checkbox is selected, select those variants that have \n",
    "        #an extreme value in at least one of the populations selected\n",
    "        if (input$extremeValuesInput == TRUE){\n",
    "            for (POP in pops){\n",
    "                if (POP %in% input$popsInput){\n",
    "                    llista_rsid <- c(llista_rsid, \n",
    "                                     filtered %>% filter(!is.na(filtered[[paste0('extremevalue_',POP)]])) %>%\n",
    "                                         select(rsid))\n",
    "                        \n",
    "                    }\n",
    "            }\n",
    "                filtered <- filtered %>% filter(rsid %in% llista_rsid)      \n",
    "        }\n",
    "        \n",
    "        #If GWAS checkbox is selected, select those variants that have a GWAS Association Count \n",
    "        #value inside the chosen range \n",
    "        if (input$filterGWASInput == T){\n",
    "            filtered <- filtered %>% filter(GWASAssocCounts >= input$GWASInput[1],\n",
    "                                            GWASAssocCounts <= input$GWASInput[2])\n",
    "        }\n",
    "        \n",
    "        #If VEP checkbox is selected, select those variants with at least one of the VEP Consequences chosen\n",
    "        if (input$filterVEPInput == T){\n",
    "            separatedVEP <- separate_rows(filtered, VEPConseqs, sep = ',')\n",
    "            ids_filter_VEP <- separatedVEP %>% filter(VEPConseqs %in% input$VEPInput)\n",
    "            filtered <- filtered %>% filter (rsid %in% ids_filter_VEP$rsid)\n",
    "        }\n",
    "        \n",
    "        #If ClinVar checkbox is selected, select those variants with at least one of the Clinical Significances chosen\n",
    "        if (input$filterClinVarInput == T){\n",
    "            separatedCV <- separate_rows(filtered, ClinicalSignificances, sep = ', ')\n",
    "            ids_filter_CV <- separatedCV %>% filter(ClinicalSignificances %in% input$ClinVarInput)\n",
    "            filtered <- filtered %>% filter (rsid %in% ids_filter_CV$rsid)\n",
    "        }\n",
    "        \n",
    "        #If Regulome checkbox is selected, select those variants with a regulome rank score value\n",
    "        #inside the chosen range\n",
    "        if (input$filterRegulomeInput == T){\n",
    "            filtered <- filtered %>% filter(RegulomeImportance >= match(input$RegulomeInput[1], regulomeLabels),\n",
    "                                            RegulomeImportance <= match(input$RegulomeInput[2], regulomeLabels))\n",
    "        }\n",
    "        \n",
    "        #If DSI checbox is selected, select those variants with a DSI value inside the chosen range\n",
    "        if (input$filterDSIInput == T){\n",
    "            filtered <- filtered %>% filter (`max(DSI)` >= input$DSIInput[1],\n",
    "                                             `max(DSI)` <= input$DSIInput[2])\n",
    "        }\n",
    "        \n",
    "        #If DPI checbox is selected, select those variants with a DPI value inside the chosen range\n",
    "        if (input$filterDPIInput == T){\n",
    "            filtered <- filtered %>% filter (`max(DPI)` >= input$DPIInput[1],\n",
    "                                             `max(DPI)` <= input$DPIInput[2])\n",
    "        }\n",
    "        \n",
    "        #If score checbox is selected, select those variants with a DisGeNET score inside the chosen range\n",
    "        if (input$filterScoreInput == T){\n",
    "            filtered <- filtered %>% filter (`max(score)` >= input$DGNscoreInput[1],\n",
    "                                             `max(score)` <= input$DGNscoreInput[2])\n",
    "        }\n",
    "        \n",
    "        #If N.of PubMed IDs checbox is selected, select those variants with a number of \n",
    "        #PubMed IDs value inside the chosen range\n",
    "        if (input$filterNofPmidsInput == T){\n",
    "            filtered <- filtered %>% filter (`max(NofPmids)` >= input$NofPmidsInput[1],\n",
    "                                             `max(NofPmids)` <= input$NofPmidsInput[2])\n",
    "        }\n",
    "        \n",
    "        #If Age filter checbox is selected, select those variants with a Age mean and score value, \n",
    "        #of the chosen clock model, inside the chosen ranges\n",
    "        if (input$filterAgeInput == T){\n",
    "            if ('Mut' %in% input$AgeModelInput){\n",
    "                filtered <- filtered %>% filter(AgeMean_Mut >= input$AgeMeanInput[1],\n",
    "                                                AgeMean_Mut <= input$AgeMeanInput[2],\n",
    "                                                QualScore_Mut >= input$qualScoreInput[1],\n",
    "                                                QualScore_Mut <= input$qualScoreInput[2])\n",
    "            }\n",
    "            if ('Rec' %in% input$AgeModelInput){\n",
    "                filtered <- filtered %>% filter(AgeMean_Rec >= input$AgeMeanInput[1],\n",
    "                                                AgeMean_Rec <= input$AgeMeanInput[2],\n",
    "                                                QualScore_Rec >= input$qualScoreInput[1],\n",
    "                                                QualScore_Rec <= input$qualScoreInput[2])\n",
    "            }\n",
    "            if ('Jnt' %in% input$AgeModelInput){\n",
    "                filtered <- filtered %>% filter(AgeMean_Jnt >= input$AgeMeanInput[1],\n",
    "                                                AgeMean_Jnt <= input$AgeMeanInput[2],\n",
    "                                                QualScore_Jnt >= input$qualScoreInput[1],\n",
    "                                                QualScore_Jnt <= input$qualScoreInput[2])\n",
    "            }\n",
    "        }\n",
    "        \n",
    "        #Design the plot\n",
    "        o <- ggplot(filtered, aes(x=RegulomeScore, stat=\"count\", fill = ..count..)) + geom_bar() \n",
    "        o + scale_fill_gradient(low = 'skyblue', high = 'purple')\n",
    "        \n",
    "    })\n",
    "    \n",
    "    \n",
    "    #DSI PLOT\n",
    "    output$plotDSI <- renderPlot({\n",
    "       \n",
    "        #Filter the original data by position and save it in filtered\n",
    "        filtered <- dades %>% filter(chr %in% input$chrInput,\n",
    "                                     physicalPos >= input$minPosInput,\n",
    "                                     physicalPos <= input$maxPosInput)\n",
    "        \n",
    "        #if extreme values checkbox is selected, select those variants that have \n",
    "        #an extreme value in at least one of the populations selected\n",
    "        if (input$extremeValuesInput == TRUE){\n",
    "            for (POP in pops){\n",
    "                if (POP %in% input$popsInput){\n",
    "                    llista_rsid <- c(llista_rsid, \n",
    "                                     filtered %>% filter(!is.na(filtered[[paste0('extremevalue_',POP)]])) %>%\n",
    "                                         select(rsid))\n",
    "                        \n",
    "                    }\n",
    "            }\n",
    "                filtered <- filtered %>% filter(rsid %in% llista_rsid)      \n",
    "        }\n",
    "        \n",
    "        #If GWAS checkbox is selected, select those variants that have a GWAS Association Count \n",
    "        #value inside the chosen range \n",
    "        if (input$filterGWASInput == T){\n",
    "            filtered <- filtered %>% filter(GWASAssocCounts >= input$GWASInput[1],\n",
    "                                            GWASAssocCounts <= input$GWASInput[2])\n",
    "        }\n",
    "        \n",
    "        #If VEP checkbox is selected, select those variants with at least one of the VEP Consequences chosen\n",
    "        if (input$filterVEPInput == T){\n",
    "            separatedVEP <- separate_rows(filtered, VEPConseqs, sep = ',')\n",
    "            ids_filter_VEP <- separatedVEP %>% filter(VEPConseqs %in% input$VEPInput)\n",
    "            filtered <- filtered %>% filter (rsid %in% ids_filter_VEP$rsid)\n",
    "        }\n",
    "        \n",
    "        #If ClinVar checkbox is selected, select those variants with at least one of the Clinical Significances chosen\n",
    "        if (input$filterClinVarInput == T){\n",
    "            separatedCV <- separate_rows(filtered, ClinicalSignificances, sep = ', ')\n",
    "            ids_filter_CV <- separatedCV %>% filter(ClinicalSignificances %in% input$ClinVarInput)\n",
    "            filtered <- filtered %>% filter (rsid %in% ids_filter_CV$rsid)\n",
    "        }\n",
    "        \n",
    "        #If Regulome checkbox is selected, select those variants with a regulome rank score value\n",
    "        #inside the chosen range\n",
    "        if (input$filterRegulomeInput == T){\n",
    "            filtered <- filtered %>% filter(RegulomeImportance >= match(input$RegulomeInput[1], regulomeLabels),\n",
    "                                            RegulomeImportance <= match(input$RegulomeInput[2], regulomeLabels))\n",
    "        }\n",
    "        \n",
    "        #If DSI checbox is selected, select those variants with a DSI value inside the chosen range\n",
    "        if (input$filterDSIInput == T){\n",
    "            filtered <- filtered %>% filter (`max(DSI)` >= input$DSIInput[1],\n",
    "                                             `max(DSI)` <= input$DSIInput[2])\n",
    "        }\n",
    "        \n",
    "        #If DPI checbox is selected, select those variants with a DPI value inside the chosen range\n",
    "        if (input$filterDPIInput == T){\n",
    "            filtered <- filtered %>% filter (`max(DPI)` >= input$DPIInput[1],\n",
    "                                             `max(DPI)` <= input$DPIInput[2])\n",
    "        }\n",
    "        \n",
    "        #If score checbox is selected, select those variants with a DisGeNET score inside the chosen range\n",
    "        if (input$filterScoreInput == T){\n",
    "            filtered <- filtered %>% filter (`max(score)` >= input$DGNscoreInput[1],\n",
    "                                             `max(score)` <= input$DGNscoreInput[2])\n",
    "        }\n",
    "        \n",
    "        #If N.of PubMed IDs checbox is selected, select those variants with a number of \n",
    "        #PubMed IDs value inside the chosen range\n",
    "        if (input$filterNofPmidsInput == T){\n",
    "            filtered <- filtered %>% filter (`max(NofPmids)` >= input$NofPmidsInput[1],\n",
    "                                             `max(NofPmids)` <= input$NofPmidsInput[2])\n",
    "        }\n",
    "        \n",
    "        #If Age filter checbox is selected, select those variants with a Age mean and score value, \n",
    "        #of the chosen clock model, inside the chosen ranges\n",
    "        if (input$filterAgeInput == T){\n",
    "            if ('Mut' %in% input$AgeModelInput){\n",
    "                filtered <- filtered %>% filter(AgeMean_Mut >= input$AgeMeanInput[1],\n",
    "                                                AgeMean_Mut <= input$AgeMeanInput[2],\n",
    "                                                QualScore_Mut >= input$qualScoreInput[1],\n",
    "                                                QualScore_Mut <= input$qualScoreInput[2])\n",
    "            }\n",
    "            if ('Rec' %in% input$AgeModelInput){\n",
    "                filtered <- filtered %>% filter(AgeMean_Rec >= input$AgeMeanInput[1],\n",
    "                                                AgeMean_Rec <= input$AgeMeanInput[2],\n",
    "                                                QualScore_Rec >= input$qualScoreInput[1],\n",
    "                                                QualScore_Rec <= input$qualScoreInput[2])\n",
    "            }\n",
    "            if ('Jnt' %in% input$AgeModelInput){\n",
    "                filtered <- filtered %>% filter(AgeMean_Jnt >= input$AgeMeanInput[1],\n",
    "                                                AgeMean_Jnt <= input$AgeMeanInput[2],\n",
    "                                                QualScore_Jnt >= input$qualScoreInput[1],\n",
    "                                                QualScore_Jnt <= input$qualScoreInput[2])\n",
    "            }\n",
    "        }\n",
    "        \n",
    "        #Design the plot\n",
    "        o <- ggplot(filtered, aes(x=`max(DSI)`, y=..count.., fill = ..count..)) + geom_histogram() \n",
    "        o + scale_fill_gradient(low='skyblue', high = 'purple')\n",
    "        \n",
    "    })\n",
    "    #DPI PLOT\n",
    "    output$plotDPI <- renderPlot({\n",
    "        \n",
    "        #Filter the original data by position and save it in filtered\n",
    "        filtered <- dades %>% filter(chr %in% input$chrInput,\n",
    "                                     physicalPos >= input$minPosInput,\n",
    "                                     physicalPos <= input$maxPosInput)\n",
    "        \n",
    "        #if extreme values checkbox is selected, select those variants that have \n",
    "        #an extreme value in at least one of the populations selected\n",
    "        if (input$extremeValuesInput == TRUE){\n",
    "            for (POP in pops){\n",
    "                if (POP %in% input$popsInput){\n",
    "                    llista_rsid <- c(llista_rsid, \n",
    "                                     filtered %>% filter(!is.na(filtered[[paste0('extremevalue_',POP)]])) %>%\n",
    "                                         select(rsid))\n",
    "                        \n",
    "                    }\n",
    "            }\n",
    "                filtered <- filtered %>% filter(rsid %in% llista_rsid)      \n",
    "        }\n",
    "        \n",
    "        #If GWAS checkbox is selected, select those variants that have a GWAS Association Count \n",
    "        #value inside the chosen range \n",
    "        if (input$filterGWASInput == T){\n",
    "            filtered <- filtered %>% filter(GWASAssocCounts >= input$GWASInput[1],\n",
    "                                            GWASAssocCounts <= input$GWASInput[2])\n",
    "        }\n",
    "        \n",
    "        #If VEP checkbox is selected, select those variants with at least one of the VEP Consequences chosen\n",
    "        if (input$filterVEPInput == T){\n",
    "            separatedVEP <- separate_rows(filtered, VEPConseqs, sep = ',')\n",
    "            ids_filter_VEP <- separatedVEP %>% filter(VEPConseqs %in% input$VEPInput)\n",
    "            filtered <- filtered %>% filter (rsid %in% ids_filter_VEP$rsid)\n",
    "        }\n",
    "        \n",
    "        #If ClinVar checkbox is selected, select those variants with at least one of the Clinical Significances chosen\n",
    "        if (input$filterClinVarInput == T){\n",
    "            separatedCV <- separate_rows(filtered, ClinicalSignificances, sep = ', ')\n",
    "            ids_filter_CV <- separatedCV %>% filter(ClinicalSignificances %in% input$ClinVarInput)\n",
    "            filtered <- filtered %>% filter (rsid %in% ids_filter_CV$rsid)\n",
    "        }\n",
    "        \n",
    "        #If Regulome checkbox is selected, select those variants with a regulome rank score value\n",
    "        #inside the chosen range\n",
    "        if (input$filterRegulomeInput == T){\n",
    "            filtered <- filtered %>% filter(RegulomeImportance >= match(input$RegulomeInput[1], regulomeLabels),\n",
    "                                            RegulomeImportance <= match(input$RegulomeInput[2], regulomeLabels))\n",
    "        }\n",
    "        \n",
    "        #If DSI checbox is selected, select those variants with a DSI value inside the chosen range\n",
    "        if (input$filterDSIInput == T){\n",
    "            filtered <- filtered %>% filter (`max(DSI)` >= input$DSIInput[1],\n",
    "                                             `max(DSI)` <= input$DSIInput[2])\n",
    "        }\n",
    "        \n",
    "        #If DPI checbox is selected, select those variants with a DPI value inside the chosen range\n",
    "        if (input$filterDPIInput == T){\n",
    "            filtered <- filtered %>% filter (`max(DPI)` >= input$DPIInput[1],\n",
    "                                             `max(DPI)` <= input$DPIInput[2])\n",
    "        }\n",
    "        \n",
    "        #If score checbox is selected, select those variants with a DisGeNET score inside the chosen range\n",
    "        if (input$filterScoreInput == T){\n",
    "            filtered <- filtered %>% filter (`max(score)` >= input$DGNscoreInput[1],\n",
    "                                             `max(score)` <= input$DGNscoreInput[2])\n",
    "        }\n",
    "        \n",
    "        #If N.of PubMed IDs checbox is selected, select those variants with a number of \n",
    "        #PubMed IDs value inside the chosen range\n",
    "        if (input$filterNofPmidsInput == T){\n",
    "            filtered <- filtered %>% filter (`max(NofPmids)` >= input$NofPmidsInput[1],\n",
    "                                             `max(NofPmids)` <= input$NofPmidsInput[2])\n",
    "        }\n",
    "        \n",
    "        #If Age filter checbox is selected, select those variants with a Age mean and score value, \n",
    "        #of the chosen clock model, inside the chosen ranges\n",
    "        if (input$filterAgeInput == T){\n",
    "            if ('Mut' %in% input$AgeModelInput){\n",
    "                filtered <- filtered %>% filter(AgeMean_Mut >= input$AgeMeanInput[1],\n",
    "                                                AgeMean_Mut <= input$AgeMeanInput[2],\n",
    "                                                QualScore_Mut >= input$qualScoreInput[1],\n",
    "                                                QualScore_Mut <= input$qualScoreInput[2])\n",
    "            }\n",
    "            if ('Rec' %in% input$AgeModelInput){\n",
    "                filtered <- filtered %>% filter(AgeMean_Rec >= input$AgeMeanInput[1],\n",
    "                                                AgeMean_Rec <= input$AgeMeanInput[2],\n",
    "                                                QualScore_Rec >= input$qualScoreInput[1],\n",
    "                                                QualScore_Rec <= input$qualScoreInput[2])\n",
    "            }\n",
    "            if ('Jnt' %in% input$AgeModelInput){\n",
    "                filtered <- filtered %>% filter(AgeMean_Jnt >= input$AgeMeanInput[1],\n",
    "                                                AgeMean_Jnt <= input$AgeMeanInput[2],\n",
    "                                                QualScore_Jnt >= input$qualScoreInput[1],\n",
    "                                                QualScore_Jnt <= input$qualScoreInput[2])\n",
    "            }\n",
    "        }\n",
    "        \n",
    "        #Design the plot\n",
    "        o <- ggplot(filtered, aes(x=`max(DPI)`, y=..count.., fill = ..count..)) + geom_histogram() \n",
    "        o + scale_fill_gradient(low='skyblue', high = 'purple')\n",
    "        \n",
    "    })\n",
    "\n",
    "    #DISGENET SCORE PLOT\n",
    "    output$plotScore <- renderPlot({\n",
    "        \n",
    "        #Filter the original data by position and save it in filtered\n",
    "        filtered <- dades %>% filter(chr %in% input$chrInput,\n",
    "                                     physicalPos >= input$minPosInput,\n",
    "                                     physicalPos <= input$maxPosInput)\n",
    "        \n",
    "        #if extreme values checkbox is selected, select those variants that have \n",
    "        #an extreme value in at least one of the populations selected\n",
    "        if (input$extremeValuesInput == TRUE){\n",
    "            for (POP in pops){\n",
    "                if (POP %in% input$popsInput){\n",
    "                    llista_rsid <- c(llista_rsid, \n",
    "                                     filtered %>% filter(!is.na(filtered[[paste0('extremevalue_',POP)]])) %>%\n",
    "                                         select(rsid))\n",
    "                        \n",
    "                    }\n",
    "            }\n",
    "                filtered <- filtered %>% filter(rsid %in% llista_rsid)      \n",
    "        }\n",
    "        \n",
    "        #If GWAS checkbox is selected, select those variants that have a GWAS Association Count \n",
    "        #value inside the chosen range \n",
    "        if (input$filterGWASInput == T){\n",
    "            filtered <- filtered %>% filter(GWASAssocCounts >= input$GWASInput[1],\n",
    "                                            GWASAssocCounts <= input$GWASInput[2])\n",
    "        }\n",
    "        \n",
    "        #If VEP checkbox is selected, select those variants with at least one of the VEP Consequences chosen\n",
    "        if (input$filterVEPInput == T){\n",
    "            separatedVEP <- separate_rows(filtered, VEPConseqs, sep = ',')\n",
    "            ids_filter_VEP <- separatedVEP %>% filter(VEPConseqs %in% input$VEPInput)\n",
    "            filtered <- filtered %>% filter (rsid %in% ids_filter_VEP$rsid)\n",
    "        }\n",
    "        \n",
    "        #If ClinVar checkbox is selected, select those variants with at least one of the Clinical Significances chosen\n",
    "        if (input$filterClinVarInput == T){\n",
    "            separatedCV <- separate_rows(filtered, ClinicalSignificances, sep = ', ')\n",
    "            ids_filter_CV <- separatedCV %>% filter(ClinicalSignificances %in% input$ClinVarInput)\n",
    "            filtered <- filtered %>% filter (rsid %in% ids_filter_CV$rsid)\n",
    "        }\n",
    "        \n",
    "        #If Regulome checkbox is selected, select those variants with a regulome rank score value\n",
    "        #inside the chosen range\n",
    "        if (input$filterRegulomeInput == T){\n",
    "            filtered <- filtered %>% filter(RegulomeImportance >= match(input$RegulomeInput[1], regulomeLabels),\n",
    "                                            RegulomeImportance <= match(input$RegulomeInput[2], regulomeLabels))\n",
    "        }\n",
    "        \n",
    "        #If DSI checbox is selected, select those variants with a DSI value inside the chosen range\n",
    "        if (input$filterDSIInput == T){\n",
    "            filtered <- filtered %>% filter (`max(DSI)` >= input$DSIInput[1],\n",
    "                                             `max(DSI)` <= input$DSIInput[2])\n",
    "        }\n",
    "        \n",
    "        #If DPI checbox is selected, select those variants with a DPI value inside the chosen range\n",
    "        if (input$filterDPIInput == T){\n",
    "            filtered <- filtered %>% filter (`max(DPI)` >= input$DPIInput[1],\n",
    "                                             `max(DPI)` <= input$DPIInput[2])\n",
    "        }\n",
    "        \n",
    "        #If score checbox is selected, select those variants with a DisGeNET score inside the chosen range\n",
    "        if (input$filterScoreInput == T){\n",
    "            filtered <- filtered %>% filter (`max(score)` >= input$DGNscoreInput[1],\n",
    "                                             `max(score)` <= input$DGNscoreInput[2])\n",
    "        }\n",
    "        \n",
    "        #If N.of PubMed IDs checbox is selected, select those variants with a number of \n",
    "        #PubMed IDs value inside the chosen range\n",
    "        if (input$filterNofPmidsInput == T){\n",
    "            filtered <- filtered %>% filter (`max(NofPmids)` >= input$NofPmidsInput[1],\n",
    "                                             `max(NofPmids)` <= input$NofPmidsInput[2])\n",
    "        }\n",
    "        \n",
    "        #If Age filter checbox is selected, select those variants with a Age mean and score value, \n",
    "        #of the chosen clock model, inside the chosen ranges\n",
    "        if (input$filterAgeInput == T){\n",
    "            if ('Mut' %in% input$AgeModelInput){\n",
    "                filtered <- filtered %>% filter(AgeMean_Mut >= input$AgeMeanInput[1],\n",
    "                                                AgeMean_Mut <= input$AgeMeanInput[2],\n",
    "                                                QualScore_Mut >= input$qualScoreInput[1],\n",
    "                                                QualScore_Mut <= input$qualScoreInput[2])\n",
    "            }\n",
    "            if ('Rec' %in% input$AgeModelInput){\n",
    "                filtered <- filtered %>% filter(AgeMean_Rec >= input$AgeMeanInput[1],\n",
    "                                                AgeMean_Rec <= input$AgeMeanInput[2],\n",
    "                                                QualScore_Rec >= input$qualScoreInput[1],\n",
    "                                                QualScore_Rec <= input$qualScoreInput[2])\n",
    "            }\n",
    "            if ('Jnt' %in% input$AgeModelInput){\n",
    "                filtered <- filtered %>% filter(AgeMean_Jnt >= input$AgeMeanInput[1],\n",
    "                                                AgeMean_Jnt <= input$AgeMeanInput[2],\n",
    "                                                QualScore_Jnt >= input$qualScoreInput[1],\n",
    "                                                QualScore_Jnt <= input$qualScoreInput[2])\n",
    "            }\n",
    "        }\n",
    "        \n",
    "        #Design the plot\n",
    "        o <- ggplot(filtered, aes(x=`max(score)`, y=..count.., fill = ..count..)) + geom_histogram() \n",
    "        o + scale_fill_gradient(low='skyblue', high = 'purple')\n",
    "        \n",
    "    })\n",
    "    \n",
    "    #NUMBER OF PUBMED IDS PLOT\n",
    "    output$plotNofPmids <- renderPlot({\n",
    "        \n",
    "        #Filter the original data by position and save it in filtered\n",
    "        filtered <- dades %>% filter(chr %in% input$chrInput,\n",
    "                                     physicalPos >= input$minPosInput,\n",
    "                                     physicalPos <= input$maxPosInput)\n",
    "        \n",
    "        #if extreme values checkbox is selected, select those variants that have \n",
    "        #an extreme value in at least one of the populations selected\n",
    "        if (input$extremeValuesInput == TRUE){\n",
    "            for (POP in pops){\n",
    "                if (POP %in% input$popsInput){\n",
    "                    llista_rsid <- c(llista_rsid, \n",
    "                                     filtered %>% filter(!is.na(filtered[[paste0('extremevalue_',POP)]])) %>%\n",
    "                                         select(rsid))\n",
    "                        \n",
    "                    }\n",
    "            }\n",
    "                filtered <- filtered %>% filter(rsid %in% llista_rsid)      \n",
    "        }\n",
    "        \n",
    "        #If GWAS checkbox is selected, select those variants that have a GWAS Association Count \n",
    "        #value inside the chosen range \n",
    "        if (input$filterGWASInput == T){\n",
    "            filtered <- filtered %>% filter(GWASAssocCounts >= input$GWASInput[1],\n",
    "                                            GWASAssocCounts <= input$GWASInput[2])\n",
    "        }\n",
    "        \n",
    "        #If VEP checkbox is selected, select those variants with at least one of the VEP Consequences chosen\n",
    "        if (input$filterVEPInput == T){\n",
    "            separatedVEP <- separate_rows(filtered, VEPConseqs, sep = ',')\n",
    "            ids_filter_VEP <- separatedVEP %>% filter(VEPConseqs %in% input$VEPInput)\n",
    "            filtered <- filtered %>% filter (rsid %in% ids_filter_VEP$rsid)\n",
    "        }\n",
    "        \n",
    "        #If ClinVar checkbox is selected, select those variants with at least one of the Clinical Significances chosen\n",
    "        if (input$filterClinVarInput == T){\n",
    "            separatedCV <- separate_rows(filtered, ClinicalSignificances, sep = ', ')\n",
    "            ids_filter_CV <- separatedCV %>% filter(ClinicalSignificances %in% input$ClinVarInput)\n",
    "            filtered <- filtered %>% filter (rsid %in% ids_filter_CV$rsid)\n",
    "        }\n",
    "        \n",
    "        #If Regulome checkbox is selected, select those variants with a regulome rank score value\n",
    "        #inside the chosen range\n",
    "        if (input$filterRegulomeInput == T){\n",
    "            filtered <- filtered %>% filter(RegulomeImportance >= match(input$RegulomeInput[1], regulomeLabels),\n",
    "                                            RegulomeImportance <= match(input$RegulomeInput[2], regulomeLabels))\n",
    "        }\n",
    "        \n",
    "        #If DSI checbox is selected, select those variants with a DSI value inside the chosen range\n",
    "        if (input$filterDSIInput == T){\n",
    "            filtered <- filtered %>% filter (`max(DSI)` >= input$DSIInput[1],\n",
    "                                             `max(DSI)` <= input$DSIInput[2])\n",
    "        }\n",
    "        \n",
    "        #If DPI checbox is selected, select those variants with a DPI value inside the chosen range\n",
    "        if (input$filterDPIInput == T){\n",
    "            filtered <- filtered %>% filter (`max(DPI)` >= input$DPIInput[1],\n",
    "                                             `max(DPI)` <= input$DPIInput[2])\n",
    "        }\n",
    "        \n",
    "        #If score checbox is selected, select those variants with a DisGeNET score inside the chosen range\n",
    "        if (input$filterScoreInput == T){\n",
    "            filtered <- filtered %>% filter (`max(score)` >= input$DGNscoreInput[1],\n",
    "                                             `max(score)` <= input$DGNscoreInput[2])\n",
    "        }\n",
    "        \n",
    "        #If N.of PubMed IDs checbox is selected, select those variants with a number of \n",
    "        #PubMed IDs value inside the chosen range\n",
    "        if (input$filterNofPmidsInput == T){\n",
    "            filtered <- filtered %>% filter (`max(NofPmids)` >= input$NofPmidsInput[1],\n",
    "                                             `max(NofPmids)` <= input$NofPmidsInput[2])\n",
    "        }\n",
    "        \n",
    "        #If Age filter checbox is selected, select those variants with a Age mean and score value, \n",
    "        #of the chosen clock model, inside the chosen ranges\n",
    "        if (input$filterAgeInput == T){\n",
    "            if ('Mut' %in% input$AgeModelInput){\n",
    "                filtered <- filtered %>% filter(AgeMean_Mut >= input$AgeMeanInput[1],\n",
    "                                                AgeMean_Mut <= input$AgeMeanInput[2],\n",
    "                                                QualScore_Mut >= input$qualScoreInput[1],\n",
    "                                                QualScore_Mut <= input$qualScoreInput[2])\n",
    "            }\n",
    "            if ('Rec' %in% input$AgeModelInput){\n",
    "                filtered <- filtered %>% filter(AgeMean_Rec >= input$AgeMeanInput[1],\n",
    "                                                AgeMean_Rec <= input$AgeMeanInput[2],\n",
    "                                                QualScore_Rec >= input$qualScoreInput[1],\n",
    "                                                QualScore_Rec <= input$qualScoreInput[2])\n",
    "            }\n",
    "            if ('Jnt' %in% input$AgeModelInput){\n",
    "                filtered <- filtered %>% filter(AgeMean_Jnt >= input$AgeMeanInput[1],\n",
    "                                                AgeMean_Jnt <= input$AgeMeanInput[2],\n",
    "                                                QualScore_Jnt >= input$qualScoreInput[1],\n",
    "                                                QualScore_Jnt <= input$qualScoreInput[2])\n",
    "            }\n",
    "        }\n",
    "        \n",
    "        #Design the plot\n",
    "        o <- ggplot(filtered, aes(x=`max(NofPmids)`, y=..count.., fill = ..count..)) + geom_histogram() \n",
    "        o + scale_fill_gradient(low='skyblue', high = 'purple')\n",
    "        \n",
    "    })\n",
    "    \n",
    "    #AGE MEAN (MUTATION CLOCK) PLOT\n",
    "    output$plotAgeMeanMut <- renderPlot({\n",
    "        \n",
    "        #Filter the original data by position and save it in filtered\n",
    "        filtered <- dades %>% filter(chr %in% input$chrInput,\n",
    "                                     physicalPos >= input$minPosInput,\n",
    "                                     physicalPos <= input$maxPosInput)\n",
    "        \n",
    "        #if extreme values checkbox is selected, select those variants that have \n",
    "        #an extreme value in at least one of the populations selected\n",
    "        if (input$extremeValuesInput == TRUE){\n",
    "            for (POP in pops){\n",
    "                if (POP %in% input$popsInput){\n",
    "                    llista_rsid <- c(llista_rsid, \n",
    "                                     filtered %>% filter(!is.na(filtered[[paste0('extremevalue_',POP)]])) %>%\n",
    "                                         select(rsid))\n",
    "                        \n",
    "                    }\n",
    "            }\n",
    "                filtered <- filtered %>% filter(rsid %in% llista_rsid)      \n",
    "        }\n",
    "        \n",
    "        #If GWAS checkbox is selected, select those variants that have a GWAS Association Count \n",
    "        #value inside the chosen range \n",
    "        if (input$filterGWASInput == T){\n",
    "            filtered <- filtered %>% filter(GWASAssocCounts >= input$GWASInput[1],\n",
    "                                            GWASAssocCounts <= input$GWASInput[2])\n",
    "        }\n",
    "        \n",
    "        #If VEP checkbox is selected, select those variants with at least one of the VEP Consequences chosen\n",
    "        if (input$filterVEPInput == T){\n",
    "            separatedVEP <- separate_rows(filtered, VEPConseqs, sep = ',')\n",
    "            ids_filter_VEP <- separatedVEP %>% filter(VEPConseqs %in% input$VEPInput)\n",
    "            filtered <- filtered %>% filter (rsid %in% ids_filter_VEP$rsid)\n",
    "        }\n",
    "        \n",
    "        #If ClinVar checkbox is selected, select those variants with at least one of the Clinical Significances chosen\n",
    "        if (input$filterClinVarInput == T){\n",
    "            separatedCV <- separate_rows(filtered, ClinicalSignificances, sep = ', ')\n",
    "            ids_filter_CV <- separatedCV %>% filter(ClinicalSignificances %in% input$ClinVarInput)\n",
    "            filtered <- filtered %>% filter (rsid %in% ids_filter_CV$rsid)\n",
    "        }\n",
    "        \n",
    "        #If Regulome checkbox is selected, select those variants with a regulome rank score value\n",
    "        #inside the chosen range\n",
    "        if (input$filterRegulomeInput == T){\n",
    "            filtered <- filtered %>% filter(RegulomeImportance >= match(input$RegulomeInput[1], regulomeLabels),\n",
    "                                            RegulomeImportance <= match(input$RegulomeInput[2], regulomeLabels))\n",
    "        }\n",
    "        \n",
    "        #If DSI checbox is selected, select those variants with a DSI value inside the chosen range\n",
    "        if (input$filterDSIInput == T){\n",
    "            filtered <- filtered %>% filter (`max(DSI)` >= input$DSIInput[1],\n",
    "                                             `max(DSI)` <= input$DSIInput[2])\n",
    "        }\n",
    "        \n",
    "        #If DPI checbox is selected, select those variants with a DPI value inside the chosen range\n",
    "        if (input$filterDPIInput == T){\n",
    "            filtered <- filtered %>% filter (`max(DPI)` >= input$DPIInput[1],\n",
    "                                             `max(DPI)` <= input$DPIInput[2])\n",
    "        }\n",
    "        \n",
    "        #If score checbox is selected, select those variants with a DisGeNET score inside the chosen range\n",
    "        if (input$filterScoreInput == T){\n",
    "            filtered <- filtered %>% filter (`max(score)` >= input$DGNscoreInput[1],\n",
    "                                             `max(score)` <= input$DGNscoreInput[2])\n",
    "        }\n",
    "        \n",
    "        #If N.of PubMed IDs checbox is selected, select those variants with a number of \n",
    "        #PubMed IDs value inside the chosen range\n",
    "        if (input$filterNofPmidsInput == T){\n",
    "            filtered <- filtered %>% filter (`max(NofPmids)` >= input$NofPmidsInput[1],\n",
    "                                             `max(NofPmids)` <= input$NofPmidsInput[2])\n",
    "        }\n",
    "        \n",
    "        #If Age filter checbox is selected, select those variants with a Age mean and score value, \n",
    "        #of the chosen clock model, inside the chosen ranges\n",
    "        if (input$filterAgeInput == T){\n",
    "            if ('Mut' %in% input$AgeModelInput){\n",
    "                filtered <- filtered %>% filter(AgeMean_Mut >= input$AgeMeanInput[1],\n",
    "                                                AgeMean_Mut <= input$AgeMeanInput[2],\n",
    "                                                QualScore_Mut >= input$qualScoreInput[1],\n",
    "                                                QualScore_Mut <= input$qualScoreInput[2])\n",
    "            }\n",
    "            if ('Rec' %in% input$AgeModelInput){\n",
    "                filtered <- filtered %>% filter(AgeMean_Rec >= input$AgeMeanInput[1],\n",
    "                                                AgeMean_Rec <= input$AgeMeanInput[2],\n",
    "                                                QualScore_Rec >= input$qualScoreInput[1],\n",
    "                                                QualScore_Rec <= input$qualScoreInput[2])\n",
    "            }\n",
    "            if ('Jnt' %in% input$AgeModelInput){\n",
    "                filtered <- filtered %>% filter(AgeMean_Jnt >= input$AgeMeanInput[1],\n",
    "                                                AgeMean_Jnt <= input$AgeMeanInput[2],\n",
    "                                                QualScore_Jnt >= input$qualScoreInput[1],\n",
    "                                                QualScore_Jnt <= input$qualScoreInput[2])\n",
    "            }\n",
    "        }\n",
    "        \n",
    "        #Design the plot\n",
    "        o <- ggplot(filtered, aes(x=AgeMean_Mut, y=..count.., fill = ..count..)) + geom_histogram() \n",
    "        o + scale_fill_gradient(low='skyblue', high = 'purple')\n",
    "        \n",
    "    })\n",
    "    \n",
    "    #AGE MEAN (RECOMBINATION CLOCK) PLOT\n",
    "    output$plotAgeMeanRec <- renderPlot({\n",
    "        \n",
    "        #Filter the original data by position and save it in filtered\n",
    "        filtered <- dades %>% filter(chr %in% input$chrInput,\n",
    "                                     physicalPos >= input$minPosInput,\n",
    "                                     physicalPos <= input$maxPosInput)\n",
    "        \n",
    "        #if extreme values checkbox is selected, select those variants that have \n",
    "        #an extreme value in at least one of the populations selected\n",
    "        if (input$extremeValuesInput == TRUE){\n",
    "            for (POP in pops){\n",
    "                if (POP %in% input$popsInput){\n",
    "                    llista_rsid <- c(llista_rsid, \n",
    "                                     filtered %>% filter(!is.na(filtered[[paste0('extremevalue_',POP)]])) %>%\n",
    "                                         select(rsid))\n",
    "                        \n",
    "                    }\n",
    "            }\n",
    "                filtered <- filtered %>% filter(rsid %in% llista_rsid)      \n",
    "        }\n",
    "        \n",
    "        #If GWAS checkbox is selected, select those variants that have a GWAS Association Count \n",
    "        #value inside the chosen range \n",
    "        if (input$filterGWASInput == T){\n",
    "            filtered <- filtered %>% filter(GWASAssocCounts >= input$GWASInput[1],\n",
    "                                            GWASAssocCounts <= input$GWASInput[2])\n",
    "        }\n",
    "        \n",
    "        #If VEP checkbox is selected, select those variants with at least one of the VEP Consequences chosen\n",
    "        if (input$filterVEPInput == T){\n",
    "            separatedVEP <- separate_rows(filtered, VEPConseqs, sep = ',')\n",
    "            ids_filter_VEP <- separatedVEP %>% filter(VEPConseqs %in% input$VEPInput)\n",
    "            filtered <- filtered %>% filter (rsid %in% ids_filter_VEP$rsid)\n",
    "        }\n",
    "        \n",
    "        #If ClinVar checkbox is selected, select those variants with at least one of the Clinical Significances chosen\n",
    "        if (input$filterClinVarInput == T){\n",
    "            separatedCV <- separate_rows(filtered, ClinicalSignificances, sep = ', ')\n",
    "            ids_filter_CV <- separatedCV %>% filter(ClinicalSignificances %in% input$ClinVarInput)\n",
    "            filtered <- filtered %>% filter (rsid %in% ids_filter_CV$rsid)\n",
    "        }\n",
    "        \n",
    "        #If Regulome checkbox is selected, select those variants with a regulome rank score value\n",
    "        #inside the chosen range\n",
    "        if (input$filterRegulomeInput == T){\n",
    "            filtered <- filtered %>% filter(RegulomeImportance >= match(input$RegulomeInput[1], regulomeLabels),\n",
    "                                            RegulomeImportance <= match(input$RegulomeInput[2], regulomeLabels))\n",
    "        }\n",
    "        \n",
    "        #If DSI checbox is selected, select those variants with a DSI value inside the chosen range\n",
    "        if (input$filterDSIInput == T){\n",
    "            filtered <- filtered %>% filter (`max(DSI)` >= input$DSIInput[1],\n",
    "                                             `max(DSI)` <= input$DSIInput[2])\n",
    "        }\n",
    "        \n",
    "        #If DPI checbox is selected, select those variants with a DPI value inside the chosen range\n",
    "        if (input$filterDPIInput == T){\n",
    "            filtered <- filtered %>% filter (`max(DPI)` >= input$DPIInput[1],\n",
    "                                             `max(DPI)` <= input$DPIInput[2])\n",
    "        }\n",
    "        \n",
    "        #If score checbox is selected, select those variants with a DisGeNET score inside the chosen range\n",
    "        if (input$filterScoreInput == T){\n",
    "            filtered <- filtered %>% filter (`max(score)` >= input$DGNscoreInput[1],\n",
    "                                             `max(score)` <= input$DGNscoreInput[2])\n",
    "        }\n",
    "        \n",
    "        #If N.of PubMed IDs checbox is selected, select those variants with a number of \n",
    "        #PubMed IDs value inside the chosen range\n",
    "        if (input$filterNofPmidsInput == T){\n",
    "            filtered <- filtered %>% filter (`max(NofPmids)` >= input$NofPmidsInput[1],\n",
    "                                             `max(NofPmids)` <= input$NofPmidsInput[2])\n",
    "        }\n",
    "        \n",
    "        #If Age filter checbox is selected, select those variants with a Age mean and score value, \n",
    "        #of the chosen clock model, inside the chosen ranges\n",
    "        if (input$filterAgeInput == T){\n",
    "            if ('Mut' %in% input$AgeModelInput){\n",
    "                filtered <- filtered %>% filter(AgeMean_Mut >= input$AgeMeanInput[1],\n",
    "                                                AgeMean_Mut <= input$AgeMeanInput[2],\n",
    "                                                QualScore_Mut >= input$qualScoreInput[1],\n",
    "                                                QualScore_Mut <= input$qualScoreInput[2])\n",
    "            }\n",
    "            if ('Rec' %in% input$AgeModelInput){\n",
    "                filtered <- filtered %>% filter(AgeMean_Rec >= input$AgeMeanInput[1],\n",
    "                                                AgeMean_Rec <= input$AgeMeanInput[2],\n",
    "                                                QualScore_Rec >= input$qualScoreInput[1],\n",
    "                                                QualScore_Rec <= input$qualScoreInput[2])\n",
    "            }\n",
    "            if ('Jnt' %in% input$AgeModelInput){\n",
    "                filtered <- filtered %>% filter(AgeMean_Jnt >= input$AgeMeanInput[1],\n",
    "                                                AgeMean_Jnt <= input$AgeMeanInput[2],\n",
    "                                                QualScore_Jnt >= input$qualScoreInput[1],\n",
    "                                                QualScore_Jnt <= input$qualScoreInput[2])\n",
    "            }\n",
    "        }\n",
    "        \n",
    "        #Design the plot\n",
    "        o <- ggplot(filtered, aes(x=AgeMean_Rec, y=..count.., fill = ..count..)) + geom_histogram() \n",
    "        o + scale_fill_gradient(low='skyblue', high = 'purple')\n",
    "        \n",
    "    })\n",
    "    \n",
    "    #AGE MEAN (JOINT CLOCK) PLOT\n",
    "    output$plotAgeMeanJnt <- renderPlot({\n",
    "        \n",
    "        #Filter the original data by position and save it in filtered\n",
    "        filtered <- dades %>% filter(chr %in% input$chrInput,\n",
    "                                     physicalPos >= input$minPosInput,\n",
    "                                     physicalPos <= input$maxPosInput)\n",
    "        \n",
    "        #if extreme values checkbox is selected, select those variants that have \n",
    "        #an extreme value in at least one of the populations selected\n",
    "        if (input$extremeValuesInput == TRUE){\n",
    "            for (POP in pops){\n",
    "                if (POP %in% input$popsInput){\n",
    "                    llista_rsid <- c(llista_rsid, \n",
    "                                     filtered %>% filter(!is.na(filtered[[paste0('extremevalue_',POP)]])) %>%\n",
    "                                         select(rsid))\n",
    "                        \n",
    "                    }\n",
    "            }\n",
    "                filtered <- filtered %>% filter(rsid %in% llista_rsid)      \n",
    "        }\n",
    "        \n",
    "        #If GWAS checkbox is selected, select those variants that have a GWAS Association Count \n",
    "        #value inside the chosen range \n",
    "        if (input$filterGWASInput == T){\n",
    "            filtered <- filtered %>% filter(GWASAssocCounts >= input$GWASInput[1],\n",
    "                                            GWASAssocCounts <= input$GWASInput[2])\n",
    "        }\n",
    "        \n",
    "        #If VEP checkbox is selected, select those variants with at least one of the VEP Consequences chosen\n",
    "        if (input$filterVEPInput == T){\n",
    "            separatedVEP <- separate_rows(filtered, VEPConseqs, sep = ',')\n",
    "            ids_filter_VEP <- separatedVEP %>% filter(VEPConseqs %in% input$VEPInput)\n",
    "            filtered <- filtered %>% filter (rsid %in% ids_filter_VEP$rsid)\n",
    "        }\n",
    "        \n",
    "        #If ClinVar checkbox is selected, select those variants with at least one of the Clinical Significances chosen\n",
    "        if (input$filterClinVarInput == T){\n",
    "            separatedCV <- separate_rows(filtered, ClinicalSignificances, sep = ', ')\n",
    "            ids_filter_CV <- separatedCV %>% filter(ClinicalSignificances %in% input$ClinVarInput)\n",
    "            filtered <- filtered %>% filter (rsid %in% ids_filter_CV$rsid)\n",
    "        }\n",
    "        \n",
    "        #If Regulome checkbox is selected, select those variants with a regulome rank score value\n",
    "        #inside the chosen range\n",
    "        if (input$filterRegulomeInput == T){\n",
    "            filtered <- filtered %>% filter(RegulomeImportance >= match(input$RegulomeInput[1], regulomeLabels),\n",
    "                                            RegulomeImportance <= match(input$RegulomeInput[2], regulomeLabels))\n",
    "        }\n",
    "        \n",
    "        #If DSI checbox is selected, select those variants with a DSI value inside the chosen range\n",
    "        if (input$filterDSIInput == T){\n",
    "            filtered <- filtered %>% filter (`max(DSI)` >= input$DSIInput[1],\n",
    "                                             `max(DSI)` <= input$DSIInput[2])\n",
    "        }\n",
    "        \n",
    "        #If DPI checbox is selected, select those variants with a DPI value inside the chosen range\n",
    "        if (input$filterDPIInput == T){\n",
    "            filtered <- filtered %>% filter (`max(DPI)` >= input$DPIInput[1],\n",
    "                                             `max(DPI)` <= input$DPIInput[2])\n",
    "        }\n",
    "        \n",
    "        #If score checbox is selected, select those variants with a DisGeNET score inside the chosen range\n",
    "        if (input$filterScoreInput == T){\n",
    "            filtered <- filtered %>% filter (`max(score)` >= input$DGNscoreInput[1],\n",
    "                                             `max(score)` <= input$DGNscoreInput[2])\n",
    "        }\n",
    "        \n",
    "        #If N.of PubMed IDs checbox is selected, select those variants with a number of \n",
    "        #PubMed IDs value inside the chosen range\n",
    "        if (input$filterNofPmidsInput == T){\n",
    "            filtered <- filtered %>% filter (`max(NofPmids)` >= input$NofPmidsInput[1],\n",
    "                                             `max(NofPmids)` <= input$NofPmidsInput[2])\n",
    "        }\n",
    "        \n",
    "        #If Age filter checbox is selected, select those variants with a Age mean and score value, \n",
    "        #of the chosen clock model, inside the chosen ranges\n",
    "        if (input$filterAgeInput == T){\n",
    "            if ('Mut' %in% input$AgeModelInput){\n",
    "                filtered <- filtered %>% filter(AgeMean_Mut >= input$AgeMeanInput[1],\n",
    "                                                AgeMean_Mut <= input$AgeMeanInput[2],\n",
    "                                                QualScore_Mut >= input$qualScoreInput[1],\n",
    "                                                QualScore_Mut <= input$qualScoreInput[2])\n",
    "            }\n",
    "            if ('Rec' %in% input$AgeModelInput){\n",
    "                filtered <- filtered %>% filter(AgeMean_Rec >= input$AgeMeanInput[1],\n",
    "                                                AgeMean_Rec <= input$AgeMeanInput[2],\n",
    "                                                QualScore_Rec >= input$qualScoreInput[1],\n",
    "                                                QualScore_Rec <= input$qualScoreInput[2])\n",
    "            }\n",
    "            if ('Jnt' %in% input$AgeModelInput){\n",
    "                filtered <- filtered %>% filter(AgeMean_Jnt >= input$AgeMeanInput[1],\n",
    "                                                AgeMean_Jnt <= input$AgeMeanInput[2],\n",
    "                                                QualScore_Jnt >= input$qualScoreInput[1],\n",
    "                                                QualScore_Jnt <= input$qualScoreInput[2])\n",
    "            }\n",
    "        }\n",
    "        \n",
    "        #Design the plot\n",
    "        o <- ggplot(filtered, aes(x=AgeMean_Jnt, y=..count.., fill = ..count..)) + geom_histogram() \n",
    "        o + scale_fill_gradient(low='skyblue', high = 'purple')\n",
    "        \n",
    "    })\n",
    "        \n",
    "        #RESULTS TABLE\n",
    "        output$tableResults <-  renderDataTable({\n",
    "            \n",
    "        #Filter the original data by position and save it in filtered\n",
    "        filtered <- dades %>% filter(chr %in% input$chrInput,\n",
    "                                     physicalPos >= input$minPosInput,\n",
    "                                     physicalPos <= input$maxPosInput)\n",
    "        \n",
    "        #if extreme values checkbox is selected, select those variants that have \n",
    "        #an extreme value in at least one of the populations selected\n",
    "        if (input$extremeValuesInput == TRUE){\n",
    "            for (POP in pops){\n",
    "                if (POP %in% input$popsInput){\n",
    "                    llista_rsid <- c(llista_rsid, \n",
    "                                     filtered %>% filter(!is.na(filtered[[paste0('extremevalue_',POP)]])) %>%\n",
    "                                         select(rsid))\n",
    "                        \n",
    "                    }\n",
    "            }\n",
    "                filtered <- filtered %>% filter(rsid %in% llista_rsid)      \n",
    "        }\n",
    "        \n",
    "        #If GWAS checkbox is selected, select those variants that have a GWAS Association Count \n",
    "        #value inside the chosen range \n",
    "        if (input$filterGWASInput == T){\n",
    "            filtered <- filtered %>% filter(GWASAssocCounts >= input$GWASInput[1],\n",
    "                                            GWASAssocCounts <= input$GWASInput[2])\n",
    "        }\n",
    "        \n",
    "        #If VEP checkbox is selected, select those variants with at least one of the VEP Consequences chosen\n",
    "        if (input$filterVEPInput == T){\n",
    "            separatedVEP <- separate_rows(filtered, VEPConseqs, sep = ',')\n",
    "            ids_filter_VEP <- separatedVEP %>% filter(VEPConseqs %in% input$VEPInput)\n",
    "            filtered <- filtered %>% filter (rsid %in% ids_filter_VEP$rsid)\n",
    "        }\n",
    "        \n",
    "        #If ClinVar checkbox is selected, select those variants with at least one of the Clinical Significances chosen\n",
    "        if (input$filterClinVarInput == T){\n",
    "            separatedCV <- separate_rows(filtered, ClinicalSignificances, sep = ', ')\n",
    "            ids_filter_CV <- separatedCV %>% filter(ClinicalSignificances %in% input$ClinVarInput)\n",
    "            filtered <- filtered %>% filter (rsid %in% ids_filter_CV$rsid)\n",
    "        }\n",
    "        \n",
    "        #If Regulome checkbox is selected, select those variants with a regulome rank score value\n",
    "        #inside the chosen range\n",
    "        if (input$filterRegulomeInput == T){\n",
    "            filtered <- filtered %>% filter(RegulomeImportance >= match(input$RegulomeInput[1], regulomeLabels),\n",
    "                                            RegulomeImportance <= match(input$RegulomeInput[2], regulomeLabels))\n",
    "        }\n",
    "        \n",
    "        #If DSI checbox is selected, select those variants with a DSI value inside the chosen range\n",
    "        if (input$filterDSIInput == T){\n",
    "            filtered <- filtered %>% filter (`max(DSI)` >= input$DSIInput[1],\n",
    "                                             `max(DSI)` <= input$DSIInput[2])\n",
    "        }\n",
    "        \n",
    "        #If DPI checbox is selected, select those variants with a DPI value inside the chosen range\n",
    "        if (input$filterDPIInput == T){\n",
    "            filtered <- filtered %>% filter (`max(DPI)` >= input$DPIInput[1],\n",
    "                                             `max(DPI)` <= input$DPIInput[2])\n",
    "        }\n",
    "        \n",
    "        #If score checbox is selected, select those variants with a DisGeNET score inside the chosen range\n",
    "        if (input$filterScoreInput == T){\n",
    "            filtered <- filtered %>% filter (`max(score)` >= input$DGNscoreInput[1],\n",
    "                                             `max(score)` <= input$DGNscoreInput[2])\n",
    "        }\n",
    "        \n",
    "        #If N.of PubMed IDs checbox is selected, select those variants with a number of \n",
    "        #PubMed IDs value inside the chosen range\n",
    "        if (input$filterNofPmidsInput == T){\n",
    "            filtered <- filtered %>% filter (`max(NofPmids)` >= input$NofPmidsInput[1],\n",
    "                                             `max(NofPmids)` <= input$NofPmidsInput[2])\n",
    "        }\n",
    "        \n",
    "        #If Age filter checbox is selected, select those variants with a Age mean and score value, \n",
    "        #of the chosen clock model, inside the chosen ranges\n",
    "        if (input$filterAgeInput == T){\n",
    "            if ('Mut' %in% input$AgeModelInput){\n",
    "                filtered <- filtered %>% filter(AgeMean_Mut >= input$AgeMeanInput[1],\n",
    "                                                AgeMean_Mut <= input$AgeMeanInput[2],\n",
    "                                                QualScore_Mut >= input$qualScoreInput[1],\n",
    "                                                QualScore_Mut <= input$qualScoreInput[2])\n",
    "            }\n",
    "            if ('Rec' %in% input$AgeModelInput){\n",
    "                filtered <- filtered %>% filter(AgeMean_Rec >= input$AgeMeanInput[1],\n",
    "                                                AgeMean_Rec <= input$AgeMeanInput[2],\n",
    "                                                QualScore_Rec >= input$qualScoreInput[1],\n",
    "                                                QualScore_Rec <= input$qualScoreInput[2])\n",
    "            }\n",
    "            if ('Jnt' %in% input$AgeModelInput){\n",
    "                filtered <- filtered %>% filter(AgeMean_Jnt >= input$AgeMeanInput[1],\n",
    "                                                AgeMean_Jnt <= input$AgeMeanInput[2],\n",
    "                                                QualScore_Jnt >= input$qualScoreInput[1],\n",
    "                                                QualScore_Jnt <= input$qualScoreInput[2])\n",
    "            }\n",
    "        }\n",
    "            #Don't select the columns made by us (numeric impacts)\n",
    "            return(filtered %>% select(-MaxNumConseq, -BinaryConseq, -ClinicalImpacts, -MaxClinicalImpact, -RegulomeImportance)) \n",
    "                  \n",
    "        })\n",
    "    \n",
    "#DOWNLOAD BUTTON <- not working yet\n",
    "        thedata <- reactive({\n",
    "            filtered <- dades %>% filter(chr %in% input$chrInput,\n",
    "                                         physicalPos >= input$minPosInput,\n",
    "                                         physicalPos <= input$maxPosInput) \n",
    "            \n",
    "            if (input$extremeValuesInput == TRUE){\n",
    "                for (POP in pops){\n",
    "                    if (POP %in% input$popsInput){\n",
    "                        llista_rsid <- c(llista_rsid, \n",
    "                                         filtered %>% filter(!is.na(filtered[[paste0('extremevalue_',POP)]])) %>%\n",
    "                                             select(rsid))\n",
    "                        \n",
    "                    }\n",
    "                }\n",
    "                filtered <- filtered %>% filter(rsid %in% llista_rsid)  \n",
    "                \n",
    "                \n",
    "            }\n",
    "            if (input$filterGWASInput == T){\n",
    "                filtered <- filtered %>% filter(GWASAssocCounts >= input$GWASInput[1],\n",
    "                                                GWASAssocCounts <= input$GWASInput[2])\n",
    "            }\n",
    "            if (input$filterVEPInput == T){\n",
    "                separatedVEP <- separate_rows(filtered, VEPConseqs, sep = ',')\n",
    "                ids_filter_VEP <- separatedVEP %>% filter(VEPConseqs %in% input$VEPInput)\n",
    "                filtered <- filtered %>% filter (rsid %in% ids_filter_VEP$rsid)\n",
    "            }\n",
    "            if (input$filterClinVarInput == T){\n",
    "                separatedCV <- separate_rows(filtered, ClinicalSignificances, sep = ', ')\n",
    "                ids_filter_CV <- separatedCV %>% filter(ClinicalSignificances %in% input$ClinVarInput)\n",
    "                filtered <- filtered %>% filter (rsid %in% ids_filter_CV$rsid)\n",
    "            }\n",
    "            if (input$filterRegulomeInput == T){\n",
    "                filtered <- filtered %>% filter(RegulomeImportance >= match(input$RegulomeInput[1], regulomeLabels),\n",
    "                                                RegulomeImportance <= match(input$RegulomeInput[2], regulomeLabels))\n",
    "            }\n",
    "            if (input$filterDSIInput == T){\n",
    "                filtered <- filtered %>% filter (`max(DSI)` >= input$DSIInput[1],\n",
    "                                                 `max(DSI)` <= input$DSIInput[2])\n",
    "            }\n",
    "            if (input$filterDPIInput == T){\n",
    "                filtered <- filtered %>% filter (`max(DPI)` >= input$DPIInput[1],\n",
    "                                                 `max(DPI)` <= input$DPIInput[2])\n",
    "            }\n",
    "            if (input$filterScoreInput == T){\n",
    "                filtered <- filtered %>% filter (`max(score)` >= input$DGNscoreInput[1],\n",
    "                                                 `max(score)` <= input$DGNscoreInput[2])\n",
    "            }\n",
    "            if (input$filterNofPmidsInput == T){\n",
    "                filtered <- filtered %>% filter (`max(NofPmids)` >= input$NofPmidsInput[1],\n",
    "                                                 `max(NofPmids)` <= input$NofPmidsInput[2])\n",
    "            }\n",
    "            if (input$filterAgeInput == T){\n",
    "                if ('Mut' %in% input$AgeModelInput){\n",
    "                    filtered <- filtered %>% filter(AgeMean_Mut >= input$AgeMeanInput[1],\n",
    "                                                    AgeMean_Mut <= input$AgeMeanInput[2],\n",
    "                                                    QualScore_Mut >= input$qualScoreInput[1],\n",
    "                                                    QualScore_Mut <= input$qualScoreInput[2])\n",
    "                }\n",
    "                if ('Rec' %in% input$AgeModelInput){\n",
    "                    filtered <- filtered %>% filter(AgeMean_Rec >= input$AgeMeanInput[1],\n",
    "                                                    AgeMean_Rec <= input$AgeMeanInput[2],\n",
    "                                                    QualScore_Rec >= input$qualScoreInput[1],\n",
    "                                                    QualScore_Rec <= input$qualScoreInput[2])\n",
    "                }\n",
    "                if ('Jnt' %in% input$AgeModelInput){\n",
    "                    filtered <- filtered %>% filter(AgeMean_Jnt >= input$AgeMeanInput[1],\n",
    "                                                    AgeMean_Jnt <= input$AgeMeanInput[2],\n",
    "                                                    QualScore_Jnt >= input$qualScoreInput[1],\n",
    "                                                    QualScore_Jnt <= input$qualScoreInput[2])\n",
    "                }\n",
    "            }\n",
    "            return(filtered)\n",
    "        })\n",
    "        output$download <- downloadHandler(\n",
    "            filename = function(){'Data.cvs'},\n",
    "            content = function(fname){\n",
    "                write.csv(thedata(), fname)\n",
    "            } #es baixara l'arxiu amb el nom de DATA\n",
    "        )\n",
    "}\n",
    "\n",
    "\n",
    "# Run the application \n",
    "shinyApp(ui = ui, server = server)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
